---
title: "Integrated relationship of nasopharyngeal airway host response and microbiome with bronchiolitis severity :dual-transcriptomic profiling"
author: "Michimasa Fujiogi; Yoshihiko Raita; Marcos Pérez-Losada; Robert J. Freishtat,; Juan C. Celedón; Jonathan M. Mansbach; Pedro A. Piedra; Zhaozhong Zhu; Carlos A. Camargo, Jr.; and Kohei Hasegawa"
date: "2/17/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

####################### 
I. SET UP 
####################### 
```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)

options(digits=5) 
#install.packages("scales") 
#set 
Sys.setlocale("LC_MESSAGES",'en_US')

#list dele
#rm(list=ls())

# Load packages -----------------------------------------------------------
library("tidyverse")
library("rlang")
library("openxlsx")
library("readxl")
library("tableone")
library("dplyr")
library(WGCNA)
library(igraph)
# The following setting is important, do not omit.
options(stringsAsFactors = FALSE);
##Options--------------------------------------------------------------------------
#alert setting
options(warn=1)
options(max.print=999999)
#alert (English)
#Sys.setlocale("LC_ALL","English")

#create df function
createdf = function( nrow, ncol, colnames = c() ){
  data.frame( matrix( vector(), nrow, ncol, dimnames = list( c(), colnames ) ) )
}


#function get best result 
get_best_result = function(sft) {
  best = which(sft$fitIndices$SFT.R.sq == max(sft$fitIndices$SFT.R.sq))
  best_result = sft$fitIndices$SFT.R.sq[best]
  print (c(best, best_result))
}

#function get best result 
get_best_result2 = function(caret_fit) {
  best = which(rownames(caret_fit$results) == rownames(caret_fit$bestTune))
  best_result = caret_fit$results[best, ]
  rownames(best_result) = NULL
  best_result
}
```

#Setwd 
#さいごにけす
```{r For Michi's PC}
Sys.getenv("HOME")
normalizePath("..")
# getwd-setwd
getwd()
setwd("/Users/mf65/Dropbox/mghdata")
dir <- getwd()

list.files(dir)
setwd("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/wgcna")
```



#=================================================================================================================
#Preparation
#=================================================================================================================
# Import dataset
```{r}
setwd("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/Revision1_nc")

# Upload module data created by WGCNA analysis
df.all <- read.table(paste0("merged_for_integration_5top_0208.csv"),header = T, sep = ",", row.names = 1) 
#df.all <-fread("merged_for_integration.csv") 

df.all$RSVRVothers <- as.factor(df.all$RSVRVothers)
df.all$RSVRVothers <- as.factor(df.all$intake_sex)

```


A) Examine associations of each omics data element with the risk of PPV use at the individual data level. 
#Host response
```{r}
# Import data
txi <- readRDS("hostsresponsedata.rds", refhook = NULL)
df_clinicaldata <- read.csv("metadata.csv", header=TRUE) #clinical dataset

ddsTxi <- DESeqDataSetFromTximport(txi,
                                   colData = df_clinicaldata,
                                   design = ~ PPVuse)
# Minimal pre-filtering (keep only rows that have at least 10 reads total) 
keep <- rowSums(counts(ddsTxi)) >= 10
ddsTxi <- ddsTxi[keep,]

# Differential expression analysis
ddsTxi <- DESeq(ddsTxi, parallel=F)
res <- results(ddsTxi)

# Data object for fig2A
data_object_fig2A <- res

# Pathway analysis
library(clusterProfiler)
library(org.Hs.eg.db)
res <- results(ddsTxi)
all.genes <- rownames(res)
all.genes.entrez <- bitr(all.genes, fromType="ENSEMBL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
original_gene_list <- res$stat
# name the vector
names(original_gene_list) <- rownames(res)
original_gene_list_df <- as.data.frame(original_gene_list)
original_gene_list_df <- data.frame(rownames(original_gene_list_df), original_gene_list_df$original_gene_list)
host_list <- left_join(original_gene_list_df, all.genes.entrez, by = c("rownames" = "ENSEMBL"))
host_list = sort(host_list, decreasing = T)


gse_result <- gseGO(geneList     = host_list, 
                    OrgDb        = org.Hs.eg.db,
                    ont          = c("BP","CC","MF"),
                    pvalueCutoff = 1,
                    minGSSize = 10,
                    maxGSSize = 500, 
                    eps = 0, 
                    pAdjustMethod = "fdr", 
                    verbose = TRUE, 
                    by = "fgsea",
                    seed = 1
                    )

#Make data_object
#Biological process
"data_object_fig2B.csv"
#Molecular process
"data_object_figS**A.csv"
#Cellular component
"data_object_figS**B.csv"

# For WGCNA
resSig <- subset(res, padj < 0.4)
resSig <- as.data.frame(resSig)
host_04 <- rownames(resSig)
df_host_vstdata <- vst(ddsTxi, blind=FALSE)
df_host_vstdata  <- as.data.frame(t(assay(df_host_vstdata)))
df_host  <- df_host_vstdata %>% dplyr::select(all_of(host_04))
```

#Microbial composition
```{r}
library(phyloseq)
#Import data
ps <- read.csv("compositiondata.csv", header=TRUE)
colnames(tax_table(ps)) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Strain")
#
sample_data(ps) <- df_clinicaldata #clinical dataset
#Filter by other taxa with not correspondence to analysis 
filterPhyla2 <- c("k__Eukaryota", "k__Chloroplast", "k__Mitochondria")
ps1 <- subset_taxa(ps1, !Kingdom %in% filterPhyla2)
ps1 <- subset_taxa(ps1, !Phylum %in% filterPhyla2)
ps1 <- subset_taxa(ps1, !Class %in% filterPhyla2)
ps1 <- subset_taxa(ps1, !Order %in% filterPhyla2)
ps1 <- subset_taxa(ps1, !Family %in% filterPhyla2)
ps1 <- subset_taxa(ps1, !Genus %in% filterPhyla2)
ps1 <- subset_taxa(ps1, !Species %in% filterPhyla2)
ps1 <- subset_taxa(ps1, !Strain %in% filterPhyla2)
ps1

# Remove OTUs whose mean value per sample is less than 1
ps2 <- filter_taxa(ps1, function(x) mean(x) > 1, TRUE)
# We can also remove taxa that are not observed more than X times in at least 10% of the samples
ps3 <- filter_taxa(ps2, function(x) sum(x > 2) > (0.1*length(x)), TRUE)
# And finally filter samples with less than 1000 reads
ps4 = prune_samples(sample_sums(ps3) >= 1000, ps3)
ps_bacteria <- subset_taxa(ps4, Kingdom=="Bacteria")
ps_Species <- phyloseq::tax_glom(ps_bacteria, "Species")
# Convert to relative abundance
ps_rel_abund = phyloseq::transform_sample_counts(ps_Species, function(x){x / sum(x)})

ps_spe_top10 <- get_top_taxa(physeq_obj = ps_rel_abund, n = 10, relative = TRUE,
                           discard_other = T, other_label = "Other")

# Data object for fig3
data_object_fig3B <- phyloseq::psmelt(ps_spe_top10)


# Poisson regression 
ps_spe <- get_top_taxa(physeq_obj = ps_Species , n = 320, relative = TRUE,
                           discard_other = T, other_label = "Other")
df_composition <- data.frame(t(data.frame(phyloseq::otu_table(ps_spe))))

ncol <- 320
table_spe_po <- createdf(ncol,4, colnames = c( "Variable", "p.value", "FDR"))

for (i in 1:ncol) {
  table_spe_po [i,1] <- names(df_composition )[i] #Variable Name
  result1 <-  glm(df_composition[,i] ~ PPVuse , family = poisson, data=df_composition )
  result2 <- summ(result1, exp = T )
  table_spe_po[i,2] <-  result2$coeftable[2,5] #Pvalue
}


# FDR
for( i in 1:ncol){
 BH <- p.adjust(table_spe_po$p.value, method="fdr")
 table_spe_po[i,3] <- round(BH[i],digits = 3)
}
table_spe_po_fdr04 <- table_spe_po  %>% filter(FDR < 0.4)
mbtx_po04 <- table_spe_po2_fdr04$Variable

# For WGCNA
df_composition <- df_composition_countdata %>% filter(rowname %in% c(mbtx_po04))
```


#Microbial function
```{r}
df_function_pre <-  read.csv("functiondata.csv", header=TRUE)

df_function_pre <- df_function_pre  %>% left_join(df_clinicaldata ,  by="Sample_ID")

# place holder
result_func <-data.frame(name = factor(), pval=double(), stat=double(), FC =double(), log2FC=double())


ncol <- ncol(df_function_pre)
for (i in names(df_function_pre)[c(3:ncol)]) {
       model=t.test(df_function_pre[[i]] ~ PPVuse, 
                     data =df_function_pre)
       FC <- (model$estimate[2]/model$estimate[1]) #base=no PPV use
       logFC <- log(model$estimate[2]/model$estimate[1],2)
       t <-data.frame(name=i, pval=model$p.value, stat =model$statistic, FC=FC, log2FC=logFC)
       result_func <- rbind(result_func,t)
}

 result_func $pval2<-format.pval(result_func$pval, eps = .00000001, digits = 10)
 result_func$p_adjust<-  result_func$pval%>%p.adjust(., method="fdr") %>% round(., digits=10)
 result_func$p_adjust <- as.numeric(result_func$p_adjust)
 result_func <- result_func[order(result_func$p_adjust),]
 result_func$p_adjust2 <- format.pval(result_func$p_adjust, eps = .001, digits = 2)

 # Data object for fig
 
 result_func
 #######################datasetじゃないこれは
ttest_func <- ttest_func %>% filter(ttest_func$p_adjust < 0.4)
ttest_func_list <- ttest_func$name

df_function <- df_vst2 %>% dplyr::select(all_of(ttest_func_list))

#pathway analysis
result_func$ENTREZID = mapIds(org.EcK12.eg.db,
                    keys=  result_func$name ,
                    column="ENTREZID",
                    keytype ="SYMBOL",
                    multiVals="first")
# For WGCNA
df_function <- df_function_pre %>% filter(!(is.na(result_func$stat)))

# Pathway analysis
## Step 1: numeric vector
function_list <- result_func[,2]
## Step 2: named vector
names(function_list) = as.character(result_func[,1])
## Step 3: decreasing order
function_list = sort(function_list_list, decreasing = TRUE)

#GSEA
gse_result_function <- clusterProfiler::gseGO(geneList = function_list,  
                       OrgDb        = org.EcK12.eg.db,
                       ont          = c("BP","CC","MF"),
                       pvalueCutoff = 1,
                       minGSSize = 10,
                       maxGSSize = 500, 
                       eps = 0, 
                       pAdjustMethod = "fdr", 
                       verbose = TRUE, 
                       by = "fgsea",
                       seed = 1
                       )

#Make data_object
#Biological process
"data_object_fig4B.csv"
#Molecular process
"data_object_figS5A.csv"
#Cellular component
"data_object_figS5B.csv"
```



B) Reduce the dimensions of the host response, microbial composition, and microbial function data for the subsequent integrated analyses.
#Host response
```{r}
#dataset = df_host
#Remove low variance 
std_dataset <- apply(df_host, 2, sd)
sd <- quantile(std_dataset , na.rm= T, c(0, 0.1, 0.5, 1))
index_low_std <- std_dataset  < sd[2]
# Subset by sd index
df_host <- df_host[, index_low_std==FALSE]

#WGCNA
# Choose a set of soft thresholding powers
powers=c(1:15) # in practice this should include powers up to 20.
# choose power based on SFT criterion
sft=pickSoftThreshold(df.host, powerVector=powers, networkType = "signed hybrid")  #"unsigned", "signed", "signed hybrid"
(bestpower <- get_best_result(sft))

#pickSoftThreshold for calculating scale free topology fitting indices R^2 corresponding to different soft thresholding powers beta.
power.host <- bestpower[1]
#Defining a weighted gene co-expression network
ADJ1=abs(WGCNA::cor(df.host, use="p"))^power.host
k=softConnectivity(datE=df.host,power=power.host)

# Plot a histogram of k and a scale free topology plot
par(mfrow = c(1, 2))
hist(k, breaks = 50)
scaleFreePlot(k)

#WGCNA
net.host <- blockwiseModules(df.host,
                       corType="bicor", 
                       networkType="signed hybrid",
                       replaceMissingAdjacencies = FALSE,
                       power = power.host,
                       maxBlockSize= 300,
                       minModuleSize= 40,
                       deepSplit = 4,
                       detectCutHeight = 0.99999,
                       mergeCutHeight = 0.1,
                       numericLabels=TRUE,
                       reassignThreshold = 0.1,
                       saveTOMs=TRUE,
                       pamStage=TRUE,
                       pamRespectsDendro= T, 
                       saveTOMFileBase="TOM",
                       cosineCorrelation = FALSE,
                       weights.x = NULL, 
                       weights.y = NULL,
                       randomSeed = 1
                       )

moduleLabelsAutomatic=net.host$colors
# Convert labels to colors for plotting
table(moduleLabelsAutomatic)
moduleColorsAutomatic.host = labels2colors(moduleLabelsAutomatic,zeroIsGrey = T, colorSeq = c("turquoise","yellow","blue","magenta","black","red","brown","green","pink") )

# Plot the dendrogram and the module colors underneath
plotDendroAndColors(net.host$dendrograms[[blocknumber]],
                    colors=datColors,
                    groupLabels=c("Module colors"),
                    dendroLabels=FALSE,
                    hang=0.03, 
                    addGuide=TRUE,
                    guideHang=0.05)

(table(moduleColorsAutomatic.host) %>% as.data.frame())

#Calculate eigengenes
MEList=moduleEigengenes(df.host,colors=colors)
MEs.host = MEList$eigengenes
MEs.host.ori = MEs.host  
# remove "grey" module
MEs.host <- MEs.host.ori  %>% dplyr::select(-"MEgrey")



# Choose a module assignment
moduleColors= moduleColorsAutomatic.host #dynamicColors
# Define numbers of genes and samples
nhost = ncol(df.host)
nSamples = nrow(df.trait) #PPVuse
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(df.host, moduleColors)$eigengenes #moduleColors
MEs.ori = orderMEs(MEs0) 
# remove "grey" module
MEs <- MEs.ori %>% dplyr::select(-"MEgrey")
modTraitCor = cor(MEs, df.trait, use = "pairwise.complete.obs", method="spearman") #spearman
modTraitP = corPvalueStudent(modTraitCor, nSamples)


#Extract all gene name from all modules
Preparation#################################
modNames = substring(names(MEs.ori), 3)
colname <- paste0("",modNames)
N <- max(table(module_df$moduleLabelsAutomatic))
Module <- length(names(MEs.ori))
outcome_eigen <- createdf( N, Module, colnames = c( colname ) )
modulelist <- paste0("",modNames)
alltrait <- "PPVuse"
#############################################
  
for (iii in alltrait) {
  xx <- as.data.frame(df.trait[,iii])
  names(xx) = iii
  geneTraitSignificance = as.data.frame(cor(df.host, xx, use = "p"));
  GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));
  names(geneTraitSignificance) = paste("GS.", names(xx), sep="");
  names(GSPvalue) = paste("p.GS.", names(xx), sep="");
  namegs <- paste0(names(geneTraitSignificance) )
for (i in modulelist) {
 column = match(i, modNames);
 moduleGenes = moduleColors==i;
 MM <- paste0("MM",i)
#geneModuleMembership
 df1<- geneModuleMembership
 df2 = df1 %>% rownames_to_column(., "name")  #rownames_to_column()  add_rownames
 df3<- df2[moduleGenes, c(1,column+1)] %>% as.data.frame()
 df4 = df3 %>% filter(abs(df3$MM) > 0.0)
 df5 <- df4$name %>% as.data.frame()
 MM.p<- paste0(df4$name)
 #metabolite significance
 df6<- geneTraitSignificance
 df7 = df6 %>% rownames_to_column(., "name") 
 df8<- df7[moduleGenes, 1:2] %>% as.data.frame()
 df9 = df8 %>% filter(abs(df8[,2]) > 0.0)
 df10 <- df9$name %>% as.data.frame()
 MM.s<- paste0(df9$name)
 #inner-merge
 eigenm <- inner_join(df5, df10, by= ".") 
 #put in outcome table
 for (ii in 1:N) {
   outcome_eigen[ii,i] <- paste0(eigenm$.)[ii]
 }
write.csv(outcome_eigen, paste0("df_transcriptome_in_allmodule.csv"))
}} 


library(clusterProfiler)
library(org.Hs.eg.db)
genedata_human <- read.csv("hostgenelist.csv") # all host gene list  

# repeat for all modules
  wgcnagene <- read_csv("modulegenelist.csv") #gene list per modules from "df_transcriptome_in_allmodule.csv"
  colnames(wgcnagene) <- "ENSEMBL"
  wgcnagene$ENSEMBL <- as.character(wgcnagene$ENSEMBL )
  wgcnagene2 <- wgcnagene$ENSEMBL
  allhumangene <- genedata_human$X
  # convert to entrez
  entrez <- bitr(wgcnagene2, fromType="ENSEMBL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
  all_hostgenes.entrez<- bitr(allhumangene, fromType="ENSEMBL", toType="ENTREZID", OrgDb="org.Hs.eg.db")  
  ########
  # GO Enrichment
  ########
  ego_result <- clusterProfiler::enrichGO(gene = entrez$ENTREZID,   
                universe = all_humangenes.entrez$ENTREZID,    
                OrgDb = org.Hs.eg.db,                
                ont =  c("BP","CC","MF"),
                pAdjustMethod = "fdr", 
                pvalueCutoff = 1, 
                minGSSize = 10, 
                maxGSSize = 500,
                readable = TRUE)                    
  wgcna_enrichment_result <- as.data.frame(ego_result)
  write.csv(wgcna_enrichment_result3, file = ("pathwat_result.csv"), row.names = FALSE)


#Top5
df_modTraitCor<- as.data.frame(modTraitCor)
top <- arrange(df_modTraitCor, desc(abs(df_modTraitCor$PPVuse)))
temp_head5  <- head(top,5)
top5 <- paste0(rownames(temp_head5) ,"" )
# Subset by top5
eigenvalue_host.top5 <- dplyr::select(MEs.host, c(Sample_ID, all_of(top5)))
```

#Microbial composition
```{r}
#dataset = df_composition
#Remove low variance 
std_dataset <- apply(df_composition, 2, sd)
sd <- quantile(std_dataset , na.rm= T, c(0, 0.1, 0.5, 1))
index_low_std <- std_dataset  < sd[2]
# Subset by sd index
df_composition <- df_composition[, index_low_std==FALSE]

#WGCNA
# Choose a set of soft thresholding powers
powers=c(1:15) # in practice this should include powers up to 20.
# choose power based on SFT criterion
sft=pickSoftThreshold(df_composition, powerVector=powers, networkType = "signed hybrid")  #"unsigned", "signed", "signed hybrid"
(bestpower <- get_best_result(sft))

#pickSoftThreshold for calculating scale free topology fitting indices R^2 corresponding to different soft thresholding powers beta.
power.composition <- bestpower[1]
#Defining a weighted gene co-expression network
ADJ1=abs(WGCNA::cor(df_composition, use="p"))^power.composition
k=softConnectivity(datE=df_composition, power=power.composition)

# Plot a histogram of k and a scale free topology plot
par(mfrow = c(1, 2))
hist(k, breaks = 50)
scaleFreePlot(k)

#WGCNA
net.composition <- blockwiseModules(df_composition,
                       corType="bicor", 
                       networkType="signed hybrid", 
                       replaceMissingAdjacencies = F,
                       power=power.mbtx,
                       minModuleSize= 3,
                       maxBlockSize= 40,
                       deepSplit = 4,
                       detectCutHeight = 0.99999, 
                       mergeCutHeight= 0.42, 
                       numericLabels=TRUE,
                       reassignThreshold = 0.1,
                       saveTOMs=TRUE,
                       pamStage=TRUE, 
                       pamRespectsDendro=TRUE, 
                       saveTOMFileBase="TOM",
                       cosineCorrelation = FALSE,
                       weights.x = NULL, 
                       weights.y = NULL,
                       randomSeed = 1
                       )
moduleLabelsAutomatic=net.composition$colors
# Convert labels to colors for plotting
table(moduleLabelsAutomatic)
# Convert labels to colors for plotting
moduleColorsAutomatic.composition = labels2colors(moduleLabelsAutomatic, zeroIsGrey = T, colorSeq = c("darkorchid","darkgreen", "darkblue","darkorange","darkred","darkturquoise","darkcyan") )

# Plot the dendrogram and the module colors underneath
plotDendroAndColors(net.composition$dendrograms[[blocknumber]],
                    colors=datColors,
                    groupLabels=c("Module colors"),
                    dendroLabels=FALSE,
                    hang=0.03, 
                    addGuide=TRUE,
                    guideHang=0.05)

(table(moduleColorsAutomatic.composition) %>% as.data.frame())

#Calculate eigengenes
MEList=moduleEigengenes(df.host,colors=colors)
MEs.composition = MEList$eigengenes
MEs.composition.ori = MEs.composition  
# remove "grey" module
MEs.composition <- MEs.composition.ori  %>% dplyr::select(-"MEgrey")



# Choose a module assignment
moduleColors= moduleColorsAutomatic.composition
# Define numbers of genes and samples
nhost = ncol(df_composition)
nSamples = nrow(df.trait) #PPVuse
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(df_composition, moduleColors)$eigengenes #moduleColors
MEs.ori = orderMEs(MEs0) 
# remove "grey" module
MEs <- MEs.ori %>% dplyr::select(-"MEgrey")
modTraitCor = cor(MEs, df.trait, use = "pairwise.complete.obs", method="spearman") #spearman
modTraitP = corPvalueStudent(modTraitCor, nSamples)


#Extract all gene name from all modules
Preparation#################################
modNames = substring(names(MEs.ori), 3)
colname <- paste0("",modNames)
N <- max(table(module_df$moduleLabelsAutomatic))
Module <- length(names(MEs.ori))
outcome_eigen <- createdf( N, Module, colnames = c( colname ) )
modulelist <- paste0("",modNames)
alltrait <- "PPVuse"
#############################################
  
for (iii in alltrait) {
  xx <- as.data.frame(df.trait[,iii])
  names(xx) = iii
  geneTraitSignificance = as.data.frame(cor(df.host, xx, use = "p"));
  GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));
  names(geneTraitSignificance) = paste("GS.", names(xx), sep="");
  names(GSPvalue) = paste("p.GS.", names(xx), sep="");
  namegs <- paste0(names(geneTraitSignificance) )
for (i in modulelist) {
 column = match(i, modNames);
 moduleGenes = moduleColors==i;
 MM <- paste0("MM",i)
 df1<- geneModuleMembership
 df2 = df1 %>% rownames_to_column(., "name")  
 df3<- df2[moduleGenes, c(1,column+1)] %>% as.data.frame()
 df4 = df3 %>% filter(abs(df3$MM) > 0.0)
 df5 <- df4$name %>% as.data.frame()
 MM.p<- paste0(df4$name)
 df6<- geneTraitSignificance
 df7 = df6 %>% rownames_to_column(., "name") 
 df8<- df7[moduleGenes, 1:2] %>% as.data.frame()
 df9 = df8 %>% filter(abs(df8[,2]) > 0.0)
 df10 <- df9$name %>% as.data.frame()
 MM.s<- paste0(df9$name)
 #inner-merge
 eigenm <- inner_join(df5, df10, by= ".") 
 for (ii in 1:N) {
   outcome_eigen[ii,i] <- paste0(eigenm$.)[ii]
 }
write.csv(outcome_eigen, paste0("df_composition_in_allmodule.csv"))
}} 



#Top5
df_modTraitCor<- as.data.frame(modTraitCor)
top <- arrange(df_modTraitCor, desc(abs(df_modTraitCor$PPVuse)))
temp_head5  <- head(top,5)
top5 <- paste0(rownames(temp_head5) ,"" )
# Subset by top5
eigenvalue_composition.top5 <- dplyr::select(MEs.composition, c(Sample_ID, all_of(top5)))
```

#Microbial function
```{r}
#Remove low variance 
std_dataset <- apply(df_function, 2, sd)
sd <- quantile(std_dataset , na.rm= T, c(0, 0.1, 0.5, 1))
index_low_std <- std_dataset  < sd[2]
# Subset by sd index
df_function <- df_function[, index_low_std==FALSE]

#WGCNA
# Choose a set of soft thresholding powers
powers=c(1:15) # in practice this should include powers up to 20.
# choose power based on SFT criterion
sft=pickSoftThreshold(df.function, powerVector=powers, networkType = "signed hybrid")  #"unsigned", "signed", "signed hybrid"
(bestpower <- get_best_result(sft))

#pickSoftThreshold for calculating scale free topology fitting indices R^2 corresponding to different soft thresholding powers beta.
power.function <- bestpower[1]
#Defining a weighted gene co-expression network
ADJ1=abs(WGCNA::cor(df.function, use="p"))^power.function
k=softConnectivity(datE=df.function,power=power.function)

# Plot a histogram of k and a scale free topology plot
par(mfrow = c(1, 2))
hist(k, breaks = 50)
scaleFreePlot(k)

#WGCNA
net.function <- WGCNA::blockwiseModules(df.function,
                       corType="bicor", 
                       networkType="signed hybrid", 
                       replaceMissingAdjacencies = FALSE,
                       power=power.function,
                       maxBlockSize= 500, 
                       minModuleSize = 5,
                       deepSplit = 4,
                       detectCutHeight = 0.99999,
                       mergeCutHeight = 0.92, 
                       numericLabels=TRUE,
                       reassignThreshold = 0.25, 
                       saveTOMs=TRUE,
                       pamStage=TRUE,
                       pamRespectsDendro=TRUE, 
                       saveTOMFileBase="TOM",
                       cosineCorrelation = FALSE,
                       weights.x = NULL, 
                       weights.y = NULL,
                       randomSeed = 1, 
                       )


moduleLabelsAutomatic=net.function$colors
# Convert labels to colors for plotting
table(moduleLabelsAutomatic)
  moduleColorsAutomatic.function = labels2colors(moduleLabelsAutomatic,zeroIsGrey = T, colorSeq = c("lightcyan3","plum2","goldenrod","khaki","royalblue","mediumpurple","lightgreen","salmon", "violetred", "darkseagreen", "maroon2" ))

# Plot the dendrogram and the module colors underneath
plotDendroAndColors(net.function$dendrograms[[blocknumber]],
                    colors=datColors,
                    groupLabels=c("Module colors"),
                    dendroLabels=FALSE,
                    hang=0.03, 
                    addGuide=TRUE,
                    guideHang=0.05)

(table(moduleColorsAutomatic.function) %>% as.data.frame())

#Calculate eigengenes
colors=moduleColorsAutomatic.function 
MEList=moduleEigengenes(df.function,colors=colors)
MEs.function = MEList$eigengenes
MEs.function.ori = MEs.function  
# remove "grey" module
MEs.function <- MEs.function.ori  %>% select(-"MEgrey")


# Choose a module assignment
moduleColors= moduleColorsAutomatic.function 
# Define numbers of genes and samples
nhost = ncol(df_function)
nSamples = nrow(df.trait) #PPVuse
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(df_function, moduleColors)$eigengenes 
MEs.ori = orderMEs(MEs0) 
# remove "grey" module
MEs <- MEs.ori %>% dplyr::select(-"MEgrey")
modTraitCor = cor(MEs, df.trait, use = "pairwise.complete.obs", method="spearman")
modTraitP = corPvalueStudent(modTraitCor, nSamples)


#Extract all gene name from all modules
Preparation#################################
modNames = substring(names(MEs.ori), 3)
colname <- paste0("",modNames)
N <- max(table(module_df$moduleLabelsAutomatic))
Module <- length(names(MEs.ori))
outcome_eigen <- createdf( N, Module, colnames = c( colname ) )
modulelist <- paste0("",modNames)
alltrait <- "PPVuse"
#############################################
  
for (iii in alltrait) {
  xx <- as.data.frame(df.trait[,iii])
  names(xx) = iii
  geneTraitSignificance = as.data.frame(cor(df.host, xx, use = "p"));
  GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));
  names(geneTraitSignificance) = paste("GS.", names(xx), sep="");
  names(GSPvalue) = paste("p.GS.", names(xx), sep="");
  namegs <- paste0(names(geneTraitSignificance) )
for (i in modulelist) {
 column = match(i, modNames);
 moduleGenes = moduleColors==i;
 MM <- paste0("MM",i)
#geneModuleMembership
 df1<- geneModuleMembership
 df2 = df1 %>% rownames_to_column(., "name")  
 df3<- df2[moduleGenes, c(1,column+1)] %>% as.data.frame()
 df4 = df3 %>% filter(abs(df3$MM) > 0.0)
 df5 <- df4$name %>% as.data.frame()
 MM.p<- paste0(df4$name)
 #metabolite significance
 df6<- geneTraitSignificance
 df7 = df6 %>% rownames_to_column(., "name") 
 df8<- df7[moduleGenes, 1:2] %>% as.data.frame()
 df9 = df8 %>% filter(abs(df8[,2]) > 0.0)
 df10 <- df9$name %>% as.data.frame()
 MM.s<- paste0(df9$name)
 #inner-merge
 eigenm <- inner_join(df5, df10, by= ".") 
 #put in outcome table
 for (ii in 1:N) {
   outcome_eigen[ii,i] <- paste0(eigenm$.)[ii]
 }
write.csv(outcome_eigen, paste0("df_metatranscriptome_in_allmodule.csv")) 
}} 

library(clusterProfiler)
library(org.EcK12.eg.db)
all_func_gene <- read.csv(file="all_func.csv") # all metatranscriptome 

# repeat for all modules
  wgcnagene <- read_csv("modulegenelist.csv") #gene list per modules from "df_transcriptome_in_allmodule.csv"
  colnames(wgcnagene) <- "ENSEMBL"
  wgcnagene$ENSEMBL <- as.character(wgcnagene$ENSEMBL )
  wgcnagene2 <- wgcnagene$ENSEMBL
  all_func_gene  <- all_func_gene$X
  # convert to entrez
  wentrez <- bitr(wgcnagene2, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.EcK12.eg.db")
  allfuncgenes.entrez <- bitr(all_func_gene2, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.EcK12.eg.db")  
  # GO Enrichment
  ego_result <- clusterProfiler::enrichGO(gene = na.omit(entrez$ENTREZID),          
                universe = na.omit(allfuncgenes.entrez$ENTREZID), 
                OrgDb = org.EcK12.eg.db,              
                ont =   c("BP","CC","MF"),                
                pAdjustMethod = "fdr", 
                pvalueCutoff = 1.0, 
                #qvalueCutoff = 1,
                minGSSize = 10, 
                maxGSSize = 500,
                readable = TRUE)                 
  wgcna_enrichment_result <- as.data.frame(ego_result)
  write.csv(wgcna_enrichment_result3, file = ("microbiome_pathway_result.csv"), row.names = FALSE)


#Top5
df_modTraitCor<- as.data.frame(modTraitCor)
top <- arrange(df_modTraitCor, desc(abs(df_modTraitCor$PPVuse)))
temp_head5  <- head(top,5)
top5 <- paste0(rownames(temp_head5) ,"" )
# Subset by top5
eigenvalue_function.top5 <- dplyr::select(MEs.function, c(Sample_ID, all_of(top5)))
```

#Merge clinical, host response module, microbial composition module, and microbial function module datasets for the subsequent integrated analyses
```{r}
#selecet variables in clinical data
df_clinicaldata2 <- df_clinicaldata %>% dplyr::select(c(age, sex, respiratoryvirus, PPVuse)) #race/ethnicity, intensive care use(add if needed)

#Merge
df_integrated<- df_clinicaldata %>%  
                left_join(eigenvalue_host.top5,    by="Sample_ID") %>% 
                left_join(eigenvalue_composition.top5, by="Sample_ID") %>%   
                left_join(eigenvalue_function.top5, by="Sample_ID")   
```


C) Determine the integrated relationships of these dual-transcriptome modules with outcomes
Main analysis (Outcome = PPV use, Covariates = age, sex, and respiratory virus)
Sensitivity analysis 1 (#Outcome = PPV use, Covariates =age, sex, race/ethnicity, and respiratory virus)
Sensitivity analysis 2 (Outcome = intensive care use, Covariates = age, sex, and respiratory virus)
Sensitivity analysis 3 (#Outcome = intensive care use, Covariates = age, sex, race/ethnicity, and respiratory virus)
```{r}
#dataset = df_integrated 

myTrainingControl <- trainControl(method  = "LOOCV", 
                                  returnResamp="all",
                                  savePredictions = TRUE, 
                                  classProbs = TRUE,
                                  summaryFunction = twoClassSummary,
                                  preProcOptions = NULL)

#Grid
train_grid.ridge <- expand.grid(alpha  = 0, lambda = seq(0, 1,  by=0.001) )

fit_ridge <- caret::train(PPVuse ~ .,  
                              data   = df_integrated, 
                              method = "glmnet", 
                              metric = "ROC",  
                              tuneGrid  = train_grid.ridge,
                              trControl = myTrainingControl
                               )

############
# Decide model to evaluate
model_ridge_final <- fit_ridge 
###########

# Check the fitted model
plot(model_ridge_final)
#function get best result 
get_best_result = function(caret_fit) {
  best = which(rownames(caret_fit$results) == rownames(caret_fit$bestTune))
  best_result = caret_fit$results[best, ]
  rownames(best_result) = NULL
  best_result
}

best_lambda <-  get_best_result2(model_ridge_final)[,2]

# best parameter
model_ridge_final$bestTune
#Result of best parameter1
(ridge_full.or <- exp(coef(model_ridge_final$finalModel, s = best_lambda )))



#Result of best parameter
mat_coeff_ridge <- coef(model_ridge_final$finalModel,
                              model_ridge_final$bestTune$lambda) 
mat_coeff_ridge <- mat_coeff_ridge[-1,] %>%   
                                        as.matrix()       
df_coeff_ridge.pre <- mat_coeff_ridge %>% as.data.frame()
df_coeff_ridge <- rownames_to_column(df_coeff_ridge.pre, var = "variable")
names(df_coeff_ridge)[ which( names(df_coeff_ridge)=="V1") ] <- "coef"
df_coeff_ridge$OR <- exp(df_coeff_ridge$coef)
df_coeff_ridge


####################################
modulename <- df_coeff_ridge[,1]
nsim <- 2000
nofv <- length(modulename)
#Create Data holder
data_boot_holder <- createdf(nsim,nofv, colnames = c( modulename))
result_or <- createdf(nofv,7, colnames = c( "Module","OR","lowCI","upCI", "Coef", "absCoef"))
######################################

#loop
set.seed(1)
data_boot <- df_models
for (i in 1:nsim)
{
   boot_rows <-sample(x=c(1:nrow(data_boot)), size=nrow(data_boot), replace=T)
   data_boot2 <-data_boot[boot_rows,]
   boot.ridge_full  <-   caret::train(PPVuse ~ .,  
                              data   = df_integrated, 
                              method = "glmnet", 
                              metric = "ROC", 
                              tuneGrid = expand.grid(lambda = best_lambda , alpha = 0),
                              trControl = myTrainingControl
                               )
   coef <- coef(boot.ridge_full$finalModel, s = best_lambda)


 for (ii in 1:nofv) {
   data_boot_holder[i,ii] <- coef[ii+1,1]
 }}


# Extract Coef(Odds Ratio) -Confidence Interval
for (i in 1:nofv) {
  result_or[i,1] <-  df_coeff_ridge$variable[i] 
  d = quantile(data_boot_holder[,i], c(0.025,0.5, 0.975),na.rm=T)
  result_or[i,2] <- exp(d[2]) #0.5CI
  result_or[i,3] <- exp(d[1]) #0.025CI
  result_or[i,4] <- exp(d[3]) #0.975CI
  result_or[i,5] <- (d[2]) #0.5CI
  result_or[i,6] <- abs(d[2]) #0.5CI
}


# Causal structure learning
```












##########################################################################
###############################FIGURE CODE################################
##########################################################################
#============================================================================
#Figure 1: Analytic flow of integrated-omics analysis
#============================================================================
Figure 1 was made by using Lucidchart (https://www.lucidchart.com/pages/)


#============================================================================
#Figure 2:  Differential gene expression analysis of host transcriptome data with regard to the use of positive pressure ventilation in infants hospitalized for bronchiolitis
#============================================================================

#A) Volcano plot
```{r }
df_fig2A <-  read.csv("data_object_fig2A.csv", header=TRUE)


# create custom key-value pairs
keyvals <- ifelse( df_fig2A$log2FoldChange < -0.58 & df_fig2A$padj <0.05 , "royalblue",
                            ifelse(df_fig2A$log2FoldChange > 0.58 & df_fig2A$padj <0.05, "red2",
                                  ifelse(abs(df_fig2A$log2FoldChange) >= 0.58 & df_fig2A$padj >= 0.05, "forestgreen",
                                      ifelse(abs(df_fig2A$log2FoldChange) < 0.58 & df_fig2A$padj < 0.05, "#FFAB40", "grey30"))))
levels(keyvals) <- c("grey30","red2","forestgreen","#FFAB40","royalblue" )


keyvals[is.na(keyvals)] <- "grey30"
names(keyvals)[keyvals == "red2"] <- "Upregulated*"
names(keyvals)[keyvals == "grey30"] <- "Not significant"
names(keyvals)[keyvals == "forestgreen"] <- "Log FC ≥|0.58|"
names(keyvals)[keyvals == "#FFAB40"] <- "FDR<0.05"
names(keyvals)[keyvals == "royalblue"] <- "Downregulated*"
  
levels(keyvals) <- c("Upregulated","Downregulated","Log FC ≥|0.58|","FDR<0.05","Not significant" )


library(EnhancedVolcano)
library(ggrepel)
fig2A <- EnhancedVolcano::EnhancedVolcano(df_fig2A,
    lab =  rownames(df_fig2A) ,               
    x = 'log2FoldChange',
    y = "padj", 
    selectLab = c(""),
    pCutoff = 0.05,
    FCcutoff = 0.58, 
    pointSize = 1,
    legendPosition = "right",
    colCustom = keyvals,
    labSize = 3, 
    title = NULL, 
    subtitle = NULL, 
    caption = NULL, 
    xlab = bquote(~Log[2]~ 'fold change'),
    ylab = bquote(~-Log[10]~ 'FDR')
    )

fig2A_2 <- fig2A + 
  ggplot2::coord_cartesian(xlim=c(-5, 5),ylim=c(0, 8)) + 
  scale_x_continuous(breaks = c(-5,-3,-1,0,1,3,5))+ 
  scale_y_continuous(breaks = c(0, 2.5, 5.0, 7.5))
```
B)	Gene set enrichment analysis plot (transcriptome)
```{r}
df_fig2B <-  read.csv("data_object_fig2B.csv", header=TRUE)

library(DT)

# Tidy the results:
df_fig2B <- df_fig2B %>%
  as_tibble() 
df_fig2B$absNES <- abs(df_fig2B$NES)

#Select top 30
df_fig2B <- df_fig2B %>% top_n(30, wt=-FDR)

#facet 0 or 1
df_fig2B$status <- ifelse(df_fig2B$NES > 0, "Up-regulated", "Down-regulated")

#Round FDR value
df_fig2B$FDR2 <- round(df_fig2B$FDR, 3)
df_fig2B$FDR2[which(df_fig2B$FDR2 ==0.000)] <- 0.001

#set color
colors=c( "#0C2C84","#7FCDBB","#C7E9B4","#C7E9B4")
color1 <- colorRampPalette(colors = colors)(1000)


#plot
fig2B <-       df_fig2B      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05","≥0.10"), 
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", colour="FDR", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), 
                               breaks = c (0.2, 0.3, 0.4, 0.5, 0.6),
                               labels = c( "0.2", "0.3", "0.4", "0.5", "0.6")) +
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                ), 
                          axis.title.y = element_text( size=13
                                                )
                          )

#spacing between each facet
fig2B_2  <- fig2B + 
                theme(panel.spacing.x=unit(2, "lines"),
                          legend.position="right"
                      ) 
```


#============================================================================
#Figure 3: Relationship of abundant microbial species with the risk of higher severity in infants hospitalized for bronchiolitis
#============================================================================
A)	Phylogenetic plot of top 20 
Figure 3 was made by using Python with GraPhlAn software (https://huttenhower.sph.harvard.edu/graphlan/)
Here is code for python
graphlan.py --dpi 300 data_object_fig3A.xml fig3A.png --external_legends --size 5
B)　Pirate plots of top 10
```{r}
library(ggplot2)
library(ggpirate)
fig3B  <- df_fig3B %>% 
                       ggplot(data = ., aes(x = PPVuse, y = Abundance)) +
                       geom_jitter(aes( color = "black"),size = 1,color = "black", alpha=0.5, height = 0, width =0.22) +
                       geom_violin(aes(color = Species, fill=PPVuse), color= "black", alpha=0.5) + 
                       geom_pirate(aes(color = "black"),bars =F,violins = F,points=F, cis=T, lines=T,
                                   lines_params = list(size = 0.1, width = 0.6, col="black"),
                                   cis_params = list(fill = "white", size = 0.2, alpha = 0.4, width = 0.4)
                       ) +
                       scale_fill_manual(values=c("#4575B4", "#D73027")) + 
                       labs(x = "", y = "\n Relative abundance \n ") +
                       facet_wrap(~ Species, scales = "fixed", nrow = 2) +
                       coord_flip(ylim=c(0, 0.8)) +
                       theme_light() +
                       theme( axis.text = element_text( size = 14 ),
                       axis.text.x = element_text( size = 12),
                       axis.title = element_text( size = 12),
                       legend.position="none",
                       strip.text = element_text(color= "black",size = 10, face = "bold.italic")) + 
                       annotate("text", label = "FDR<0.001", size = 5, x = 2.5, y = 0.65) 
```



#============================================================================
#Figure 4: Differential gene expression analysis of microbial function data with regard to the use of positive pressure ventilation in infants hospitalized for bronchiolitis
#============================================================================
A)	Volcano plot (metatranscriptome)
```{r}
#load data
df_fig4A <- read.csv("data_object_fig4A.csv")

rownames(df_fig4A) <- df_fig4A$name
#df_volcano$p_adjust[which( df_volcano$p_adjust == 0.0000 ) ] <- 0.0001
#head(df_volcano ,10)


#Add label
keyvals <- ifelse( df_fig4A$log2FC < -.58 & df_fig4A$p_adjust <0.05 , "royalblue",
                            ifelse(df_fig4A$log2FC > 0.58 & df_fig4A$p_adjust <0.05, "red2",
                                  ifelse(abs(df_fig4A$log2FC) >= 0.58 & df_fig4A$p_adjust >= 0.05, "forestgreen",
                                      ifelse(abs(df_fig4A$log2FC) < 0.58 & df_fig4A$p_adjust < 0.05, "#FFAB40", "grey30"))))

levels(keyvals) <- c("grey30","red2","forestgreen","#FFAB40","royalblue" )


keyvals[is.na(keyvals)] <- "grey30"
names(keyvals)[keyvals == "red2"] <- "Upregulated*"
names(keyvals)[keyvals == "grey30"] <- "Not significant"
names(keyvals)[keyvals == "forestgreen"] <- "Log FC ≥|0.58|"
names(keyvals)[keyvals == "#FFAB40"] <- "FDR<0.05"
names(keyvals)[keyvals == "royalblue"] <- "Downregulated*"
  
levels(keyvals) <- c("Upregulated","Downregulated","Log FC ≥|0.58|","FDR<0.05","Not significant" )


#volcano
library(EnhancedVolcano)
library(ggrepel)

fig4A <- EnhancedVolcano::EnhancedVolcano(df_fig4A,
    lab =  rownames(df_fig4A) ,   
    x = 'log2FC',
    y = "p_adjust", 
    selectLab = c(""),
    pCutoff = 0.05,
    FCcutoff = 0.58, 
    pointSize = 1,
    legendPosition = "right",
    colCustom = keyvals,
    labSize = 3, 
    title = NULL, 
    subtitle = NULL, 
    caption = NULL, 
    xlab = bquote(~Log[2]~ 'fold change'),
    ylab = bquote(~-Log[10]~ 'FDR'),
    )

#change label
fig4A <- fig4A + 
  ggplot2::coord_cartesian(xlim=c(-5, 5),ylim=c(0, 8)) + 
  scale_x_continuous(breaks = c(-5,-3,-1,0,1,3,5))+ 
  scale_y_continuous(breaks = c(0, 2.5, 5.0, 7.5))
```
B)	Gene set enrichment analysis plot (metatranscriptome)
```{r}
df_fig4B <- read.csv("data_object_fig4B.csv")

color2 = c( "#C34C17","#ED5000","#ED5000","#ED5000","#EE6315","#EF762A","#F08A3F","#F19D54","#F3B069","#F4C47E","#F5D793")

fig_4B <-       df_fig4B      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                    scale_colour_gradientn(colours = color2, 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05", "≥0.10"),
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", colour="FDR", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), 
                               breaks = c (0.2, 0.3, 0.4, 0.5, 0.6),
                               labels = c( "0.2", "0.3", "0.4", "0.5", "0.6")) +
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13 
                                                ), 
                          axis.title.y = element_text( size=13
                                                )
                          )
#spacing between each facet
fig_4B  <- fig_4B + 
                theme(panel.spacing.x=unit(2, "lines"),
                      legend.position="right",
                      legend.text = element_text(size=10)
                      ) 
```



#============================================================================
#Figure 5: Integrated associations of the dual-transcriptome modules with the use of positive pressure ventilation in infants hospitalized for bronchiolitis
#============================================================================
A) Heatmap
```{r}
library("corrplot")
library("ggplot2")

#data matrix
df_fig5A <- read.csv("data_object_fig5A.csv")
df_fig5A <- df_fig5A[, 1:5]
rownames(df_fig5A)  <-  df_fig5A[,1]
df_fig5A <-  df_fig5A %>% dplyr::select(-"name")


col <- colorRampPalette(c("#4575B4","#74ADD1","#ABD9E9","#E0F3F8","#FEE090","#FDAE61","#F46D43","#D73027"))

ppi=300
par(mar = c(5, 1, 1, 1)); 
tiff(paste0 ("fig5A.tiff"), height = 13*ppi, width = 13*ppi, res=ppi)
###########
# Plot
corrplot(as.matrix(df_fig5A), 
         tl.cex=1.5,
         tl.srt=60,
         method = "circle",
         win.asp = 1.0,
         is.corr = FALSE,
         #cl.lim=c(-0.055, 0.055),
         col = col(100) , 
         tl.pos = "lt", 
         tl.col = c("black"),
         cl.pos = "n",
         bg="white",
         shade.lwd =5  
         )

# Add legend
colorlegend(
  colbar = col(100),
  labels = c(-0.055, -0.25, 0, 0.25, 0.055),
  at = NULL,
  xlim = c(5.5, 6.5),
  ylim = c(6, 10),
  vertical = TRUE,
  ratio.colbar = 0.4,
  lim.segment = "auto",
  align = "l",
  addlabels = TRUE#,
  #main.cex = 10,
) 

# Add color bar name 
legend(4.5,11, 
  legend = c("Eigenvalue"), 
  pch =NA, 
  bty = "n", 
  pt.cex = 3, 
  cex = 1.5, 
  text.col = "black", 
  horiz = F , 
  inset = c(0.1, 0.1))
###########
#Add line
segments(2.5, 0.5, 2.5, 15.5,col="black", lty=3,  lwd=2) 
segments(0.5, 5.5, 4.5, 5.5,col="black", lty=3,  lwd=2) 
segments(0.5, 10.5, 4.5, 10.5,col="black", lty=3,  lwd=2) 
###########
#Add *
legend(0.2 ,15.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(0.2 ,14.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(0.2 ,13.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(0.2 ,12.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(0.2 ,11.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(0.2 ,10.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(0.2 ,5.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(0.2 ,4.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(0.2 ,3.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(0.2 ,2.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(0.2 ,1.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(2.2 ,15.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(2.2 ,14.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(2.2 ,13.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(2.2 ,12.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(2.2 ,5.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
legend(2.2 ,1.7, legend = c("*"), pch =NA, bty = "n", cex = 2.5, text.col = "black",horiz = F )
dev.off()
```
B) Lollipop plot
```{r}
#load dataset
df_fig5B <- read.csv("data_object_fig5B.csv")
df_fig5B$col <- as.factor(df_fig5B$col)


df_fig5B$Module_f <- factor(df_fig5B$Module, levels=c("NADH dehydrogenase module" , "Oxidative stress response module"    ,"BCAA metabolism module" ,"mRNA metabolism module","Plasma membrane module","Haemophilus module" ,"Streptcoccus module" , "Moraxella module"  ,"MC-1 module"   ,"S. pneumoniae/S. aureus module","HR-1 module" , "Type I IFN module","GPCR module" ,"Neutrophi/IL-1 module"   ,"T cell regulation module" ))


fig_5B <- 
  df_fig5B %>%
  ggplot(aes(Module_f,as.numeric(OR))) +
  geom_segment( aes(x=Module_f, xend=Module_f, y=1, yend=as.numeric(OR), color= col)) +
  geom_point( aes(color = col, size=as.numeric(absCoef)), fill=alpha("black", 1), shape=21, stroke=0) +
  theme_light()  +
  scale_color_manual(values=c("#4575B4", "#D73027"), 
                     labels = c("Negative", "Positive"), 
                     name = "Module") +
  scale_fill_manual(values=c("#4575B4", "#D73027")) +
  theme(legend.position = "bottom",
                                panel.grid.major.x  = element_line(colour = alpha("black", 0.22)),
                                panel.grid.minor.x  =element_blank(),
                                panel.grid.major.y   =element_blank(),
                                panel.grid.minor.y   =element_blank(),
                                panel.background = element_blank(),
                                text = element_text(size=9),
                                plot.caption = element_text(hjust = 0, face= "plain"),
                                axis.text.x = element_text(face= "plain", colour = "black", size=9),
                                axis.text.y = element_text(face= "plain", colour = "black", size=9, hjust=0),
                                axis.title.x = element_text(size=9),
                                axis.title.y = element_text(size=9),
                                legend.title = element_text(colour="black", size=9,  face="plain"),
                                legend.text = element_text(colour="black", size=9, face="plain"),
                                legend.background = element_rect(fill="grey90", colour = "white"),
                                panel.border = element_blank(),
                                axis.ticks.x = element_blank())+
  geom_hline(yintercept = 1, linetype = 1, color = "lightgray") +
  coord_flip() +
  labs(x=" ",y="Odds ratio for PPV use", title="")   # + guides(size = FALSE, color =F, alpha=F, col=F)


(fig_5B_2 <- fig_5B    
            + scale_y_log10(limits = c(0.2,5.0), breaks = c(0.2,0.4,0.6,1.0,2.0,3.0,4.0,5.0)) 
            + annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 0.5, xmax = 5.5, fill = "grey80", alpha = 0.2)
            + annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 10.5, xmax = 15.7, fill = "grey80", alpha = 0.2)
            + geom_point( aes(color = col, size=as.numeric(absCoef), fill=col, alpha=0.8), shape=21, stroke=1)          
            + guides(size = "none", color ="none", alpha="none", fill= "none"))
```
C) Causal structural learning
Figure 5C  was made by using Lucidchart (https://www.lucidchart.com/pages/)



#============================================================================
Supplementary Figure 1: Study flow diagram
#============================================================================
Supplementary Figure 1 was made by using Lucidchart (https://www.lucidchart.com/pages/)




#============================================================================
Supplementary Figure 2: Gene set enrichment analysis of host transcriptome data with regard to the use of positive pressure ventilation in infants hospitalized for bronchiolitis
#============================================================================
A) GO molecular function
```{r}
library("ggplot2")
#load dataset
df_figS2A <- read.csv("data_object_figS2A.csv")


colors=c( "#0C2C84","#7FCDBB","#C7E9B4","#C7E9B4")
color1 <- colorRampPalette(colors = colors)(1000)

fig_S2A  <-      df_figS2A      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05","≥0.10"), 
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", colour="FDR", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), 
                               breaks = c (0.2, 0.3, 0.4, 0.5, 0.6),
                               labels = c( "0.2", "0.3", "0.4", "0.5", "0.6")) +
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                ), 
                          axis.title.y = element_text( size=13
                                                ) 
                          )
#spacing between each facet
fig_S2A  <- fig_S2A + 
                theme(panel.spacing.x=unit(2, "lines"),
                          legend.position="right"
                      ) 
```
B) GO cellular component
```{r}
#laod dataset
df_figS2B <- read.csv("data_object_figS2B.csv")

colors=c( "#0C2C84","#7FCDBB","#C7E9B4","#C7E9B4")
color1 <- colorRampPalette(colors = colors)(1000)

fig_S2B <-       df_figS2B      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                     scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05","≥0.10"), 
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", size="Gene ratio", colour="FDR") + 
                    scale_size_continuous(range = c(3, 10),
                               breaks = c (0.3, 0.4, 0.5, 0.6, 0.7),
                               labels = c( "0.3", "0.4", "0.5", "0.6", "0.7")) +
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                ), 
                          axis.title.y = element_text( size=13
                                                )
                          )
#spacing between each facet
fig_S2B  <- fig_S2B + 
                theme(panel.spacing.x=unit(2, "lines"),
                          legend.position="right"
                      ) 
```




#============================================================================
Supplementary Figure 3: Relationship of abundant fungal species with the risk of higher severity in infants hospitalized for bronchiolitis
#============================================================================
```{r}
library(ggpirate)
df_figS3 <- read.csv("data_object_figS3.csv")
df_figS3$Species_f <- factor(df_figS3$Species, levels=c("Malassezia restricta", 
                                                        "Malassezia globosa", 
                                                        "Sugiyamaella lignohabitans", 
                                                        "Ramularia collo-cygni",  
                                                        "Aspergillus heteromorphus", 
                                                        "Melampsora larici-populina", 
                                                        "Trametes versicolor",
                                                        "Pseudogymnoascus destructans",
                                                        "Phycomyces blakesleeanus",
                                                        "Penicillium rubens"))


figS3  <-  df_figS3 %>% 
                       ggplot(data = ., aes(x = PPVuse, y = Abundance)) +
                       geom_jitter(aes( color = "black"),size = 1,color = "black", alpha=0.5, height = 0, width =0.22) +
                       geom_violin(aes(color = Species, fill=CPAPintubate2), color= "black", alpha=0.5) + 
                       geom_pirate(aes(color = "black"),bars =F,violins = F, points=F, cis=T,lines=T,
                                   lines_params = list(size = 0.1, width = 0.6, col="black"),
                                   cis_params = list(fill = "white", size = 0.2, alpha = 0.4, width = 0.4)
                       ) +
                       scale_fill_manual(values=c("#4575B4", "#D73027")) + 
                       labs(x = "", y = "\n Relative abundance \n ") +
                       facet_wrap(~ Species_f, scales = "fixed", nrow = 2) +
                       coord_flip(ylim=c(0, 1.0 )) +
                       theme_light() +
                       theme( axis.text = element_text( size = 14 ),
                       axis.text.x = element_text( size = 12),
                       axis.title = element_text( size = 12),
                       legend.position="none",
                       strip.text = element_text(color= "black",size = 10, face = "bold.italic")) 

labels <- data.frame(Species_f=c("Malassezia restricta", 
                                 "Malassezia globosa", 
                                 "Sugiyamaella lignohabitans", 
                                 "Ramularia collo-cygni",  
                                 "Aspergillus heteromorphus", 
                                 "Melampsora larici-populina", 
                                 "Trametes versicolor",
                                 "Pseudogymnoascus destructans",
                                 "Phycomyces blakesleeanus",
                                 "Penicillium rubens"), 
                     label=c("FDR<0.001", "FDR<0.001", "FDR<0.001","FDR<0.001", "FDR<0.001",
                             "FDR<0.001","FDR<0.001", "FDR=0.10", "FDR<0.001","FDR<0.001"))


labels$Species_f <- factor(labels$Species_f, levels=c("Malassezia restricta", 
                                                      "Malassezia globosa", 
                                                      "Sugiyamaella lignohabitans", 
                                                      "Ramularia collo-cygni",  
                                                      "Aspergillus heteromorphus", 
                                                      "Melampsora larici-populina", 
                                                      "Trametes versicolor",
                                                      "Pseudogymnoascus destructans",
                                                      "Phycomyces blakesleeanus",
                                                      "Penicillium rubens" ))

figS3_2 <- figS3  +
  geom_text(data = labels, aes(label=label, size= 4.5), x = 2.5, y = 0.80, inherit.aes = F) 
```



#============================================================================
Supplementary Figure 4: Gene set enrichment analysis of microbial function data with regard to the use of positive pressure ventilation in infants hospitalized for bronchiolitis
#============================================================================
A) GO molecular function
```{r}
#load dataset
df_figS5A <- read.csv("data_object_figS5A.csv")

color2 = c( "#C34C17","#ED5000","#ED5000","#ED5000","#EE6315","#EF762A","#F08A3F","#F19D54","#F3B069","#F4C47E","#F5D793")

fig_S5A <-      df_figS5A      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05", "≥0.10"),
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", colour="FDR", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), 
                               breaks = c (0.2, 0.3, 0.4, 0.5, 0.6),
                               labels = c( "0.2", "0.3", "0.4", "0.5", "0.6")) +
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                ), 
                          axis.title.y = element_text( size=13
                                                ) )

#spacing between each facet
fig_S5A <- fig_S5A + 
                theme(panel.spacing.x=unit(2, "lines"),
                      legend.position="right",
                      legend.text = element_text(size=10)#,
                      #legend.box="vertical"
                      ) 
```
B) GO cellular component
```{r}
#load dataset
df_figS4B <- read.csv("data_object_figS4B.csv")

color2 = c( "#C34C17","#ED5000","#ED5000","#ED5000","#EE6315","#EF762A","#F08A3F","#F19D54","#F3B069","#F4C47E","#F5D793")

fig_S5B      <-     df_figS5B      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                    scale_colour_gradientn(colours = color2 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05", "≥0.10"),
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", colour="FDR", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10),  
                               breaks = c (0.2, 0.3, 0.4, 0.5, 0.6),
                               labels = c( "0.2", "0.3", "0.4", "0.5", "0.6")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                ), 
                          axis.title.y = element_text( size=13
                                                ))
#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                      legend.position="right",
                      legend.text = element_text(size=10)
                      ) 
```



#============================================================================
Supplementary Figure 5: Sensitivity analysis: Integrated relationship of the dual-transcriptome modules with the risk of intensive care use in infants hospitalized for bronchiolitis
#============================================================================
#Lollipop plot
```{r}
df_figS5 <- read.csv("data_object_figS5.csv")
df_figS5$col <- as.factor(df_figS5$col)

df_figS5$Module_f <- factor(df_figS5$Module, levels=c("NADH dehydrogenase module" , "Oxidative stress response module"    ,"BCAA metabolism module" ,"mRNA metabolism module","Plasma membrane module","Haemophilus module" ,"Streptcoccus module" , "Moraxella module"  ,"MC-1 module"   ,"S. pneumoniae/S. aureus module","HR-1 module" , "Type I IFN module","GPCR module" ,"Neutrophi/IL-1 module" ,"T cell regulation module" ))

fig_S5 <- 
  df_figS5 %>%
  ggplot(aes(Module_f, as.numeric(OR))) +
  geom_segment( aes(x=Module_f, xend=Module_f, y=1, yend=as.numeric(OR), color= col)) +
  geom_point( aes(color = col, size=as.numeric(absCoef)), fill=alpha("black", 1), shape=21, stroke=0) +
  theme_light()  +
  scale_color_manual(values=c("#4575B4", "#D73027"), 
                     labels = c("Negative", "Positive"), 
                     name = "Module") + 
  scale_fill_manual(values=c("#4575B4", "#D73027")) +
  theme(legend.position = "bottom",
                                panel.grid.major.x  = element_line(colour = alpha("black", 0.22)),
                                panel.grid.minor.x  =element_blank(),
                                panel.grid.major.y   =element_blank(),
                                panel.grid.minor.y   =element_blank(),
                                panel.background = element_blank(),
                                text = element_text(size=9),
                                plot.caption = element_text(hjust = 0, face= "plain"),
                                axis.text.x = element_text(face= "plain", colour = "black", size=9),
                                axis.text.y = element_text(face= "plain", colour = "black", size=9, hjust=0),
                                axis.title.x = element_text(size=9),
                                axis.title.y = element_text(size=9),
                                legend.title = element_text(colour="black", size=9,  face="plain"),
                                legend.text = element_text(colour="black", size=9, face="plain"),
                                legend.background = element_rect(fill="grey90", colour = "white"),
                                panel.border = element_blank(),
                                axis.ticks.x = element_blank())+
  geom_hline(yintercept = 1, linetype = 1, color = "lightgray") +
  coord_flip() +
  labs(x=" ",y="Odds ratio for intensive care use", title="") 

(fig_S5_2<- fig_S5   
            + scale_y_log10(limits = c(0.2,8.0), breaks = c(0.2,0.4,0.6,1.0,2.0,4.0,8.0)) 
            + annotate(geom = "rect", ymin = 0.2, ymax = 8, xmin = 0.5, xmax = 5.5, fill = "grey80", alpha = 0.2)
            + annotate(geom = "rect", ymin = 0.2, ymax = 8, xmin = 10.5, xmax = 15.7, fill = "grey80", alpha = 0.2)
            + geom_point( aes(color = col, size=as.numeric(absCoef), fill=col, alpha=0.8), shape=21, stroke=1)          
            + guides(size = "none", color ="none", alpha="none", fill= "none"))
```



#============================================================================
Supplementary Figure 6. Sensitivity analysis adjusting for age, sex, race/ethnicity, and virus: Integrated relationship of the dual-transcriptome modules with the risk of positive pressure ventilation use in infants hospitalized for bronchiolitis
#============================================================================
#Lollipop plot
```{r}
#load dataset
df_figS6 <- read.csv("data_object_figS6.csv")
df_figS6$col <- as.factor(df_figS6$col)

df_figS6$Module_f <- factor(df_figS6$Module, levels=c("NADH dehydrogenase module" , "Oxidative stress response module"    ,"BCAA metabolism module","mRNA metabolism module","Plasma membrane module","Haemophilus module" ,"Streptcoccus module", "Moraxella module","MC-1 module","S. pneumoniae/S. aureus module","HR-1 module", "Type I IFN module","GPCR module","Neutrophi/IL-1 module","T cell regulation module" ))


fig_S6 <- 
  df_figS6 %>%
  ggplot(aes(Module_f,as.numeric(OR))) +
  geom_segment( aes(x=Module_f, xend=Module_f, y=1, yend=as.numeric(OR), color= col)) +
  geom_point( aes(color = col, size=as.numeric(absCoef)), fill=alpha("black", 1), shape=21, stroke=0) +
  theme_light()  +
  scale_color_manual(values=c("#4575B4", "#D73027"), 
                     labels = c("Negative", "Positive"), 
                     name = "Module") + 
  scale_fill_manual(values=c("#4575B4", "#D73027")) +
  theme(legend.position = "bottom",
                                panel.grid.major.x  = element_line(colour = alpha("black", 0.22)),
                                panel.grid.minor.x  =element_blank(),
                                panel.grid.major.y   =element_blank(),
                                panel.grid.minor.y   =element_blank(),
                                panel.background = element_blank(),
                                text = element_text(size=9),
                                plot.caption = element_text(hjust = 0, face= "plain"),
                                axis.text.x = element_text(face= "plain", colour = "black", size=9),
                                axis.text.y = element_text(face= "plain", colour = "black", size=9, hjust=0),
                                axis.title.x = element_text(size=9),
                                axis.title.y = element_text(size=9),
                                legend.title = element_text(colour="black", size=9,  face="plain"),
                                legend.text = element_text(colour="black", size=9, face="plain"),
                                legend.background = element_rect(fill="grey90", colour = "white"),
                                panel.border = element_blank(),
                                axis.ticks.x = element_blank())+
  geom_hline(yintercept = 1, linetype = 1, color = "lightgray") +
  coord_flip() +
  labs(x=" ",y="Odds ratio for PPV use", title="")  

(fig_S6_2 <- fig_S6     
            + scale_y_log10(limits = c(0.2,5.0), breaks = c(0.2,0.4,0.6,1.0,2.0,3.0,4.0,5.0)) 
            + annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 0.5, xmax = 5.5, fill = "grey80", alpha = 0.2)
            + annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 10.5, xmax = 15.7, fill = "grey80", alpha = 0.2)
            + geom_point( aes(color = col, size=as.numeric(absCoef), fill=col, alpha=0.8), shape=21, stroke=1)          
            + guides(size = "none", color ="none", alpha="none", fill= "none"))
```



#============================================================================
Supplementary Figure 7. Sensitivity analysis adjusting for age, sex, race/ethnicity, and virus: Integrated relationship of the dual-transcriptome modules with the risk of intensive care use in infants hospitalized for bronchiolitis
#============================================================================
#Lollipop plot
```{r}
df_figS7 <- read.csv("data_object_figS7.csv")
df_figS7$col <- as.factor(df_figS7$col)


df_figS7$Module_f <- factor(df_figS7$Module, levels=c("NADH dehydrogenase module", "Oxidative stress response module","BCAA metabolism module" ,"mRNA metabolism module","Plasma membrane module","Haemophilus module" ,"Streptcoccus module" , "Moraxella module","MC-1 module","S. pneumoniae/S. aureus module","HF-1 module", "Type I IFN module","GPCR module" ,"Neutrophi/IL-1 module","T cell regulation module" ))


fig_S7 <- 
  df_figS7 %>%
  ggplot(aes(Module_f,as.numeric(OR))) +
  geom_segment( aes(x=Module_f, xend=Module_f, y=1, yend=as.numeric(OR), color= col)) +
  geom_point( aes(color = col, size=as.numeric(absCoef)), fill=alpha("black", 1), shape=21, stroke=0) +
  theme_light()  +
  scale_color_manual(values=c("#4575B4", "#D73027"), 
                     labels = c("Negative", "Positive"), 
                     name = "Module") + 
  scale_fill_manual(values=c("#4575B4", "#D73027")) +
  theme(legend.position = "bottom",
                                panel.grid.major.x  = element_line(colour = alpha("black", 0.22)),
                                panel.grid.minor.x  =element_blank(),
                                panel.grid.major.y   =element_blank(),
                                panel.grid.minor.y   =element_blank(),
                                panel.background = element_blank(),
                                text = element_text(size=9),
                                plot.caption = element_text(hjust = 0, face= "plain"),
                                axis.text.x = element_text(face= "plain", colour = "black", size=9),
                                axis.text.y = element_text(face= "plain", colour = "black", size=9, hjust=0),
                                axis.title.x = element_text(size=9),
                                axis.title.y = element_text(size=9),
                                legend.title = element_text(colour="black", size=9,  face="plain"),
                                legend.text = element_text(colour="black", size=9, face="plain"),
                                legend.background = element_rect(fill="grey90", colour = "white"),
                                panel.border = element_blank(),
                                axis.ticks.x = element_blank())+
  geom_hline(yintercept = 1, linetype = 1, color = "lightgray") +
  coord_flip() +
  labs(x=" ",y="Odds ratio for intensive care use", title="")   



(fig_S7_2<- fig_S7    
            + scale_y_log10(limits = c(0.2,8.0), breaks = c(0.2,0.4,0.6,1.0,2.0,4.0,8.0)) 
            + annotate(geom = "rect", ymin = 0.2, ymax = 8, xmin = 0.5, xmax = 5.5, fill = "grey80", alpha = 0.2)
            + annotate(geom = "rect", ymin = 0.2, ymax = 8, xmin = 10.5, xmax = 15.7, fill = "grey80", alpha = 0.2)
            + geom_point( aes(color = col, size=as.numeric(absCoef), fill=col, alpha=0.8), shape=21, stroke=1)          
            + guides(size = "none", color ="none", alpha="none", fill= "none"))
```



#============================================================================
Supplementary Figure 8: Correlation network among all dual-transcriptome modules and clinical variables
#============================================================================
Supplementary Figure 8 was made by using Cytoscape (https://cytoscape.org/)
Dataset= "data_object_figS4.csv"



#============================================================================
Supplementary Figure 9: Log-log plot of whole-network connectivity distribution of host response, microbial composition, and microbial function
#============================================================================
```{r}
#Host
loglogplot_host <- read.csv("data_object_figS9_HR.csv")
loglogplot_host <- as_vector(loglogplot_host)
# Plot a scale free topology plot
WGCNA::scaleFreePlot(loglogplot_host)

#Microbiome composition
loglogplot_composition <- read.csv("data_object_figS9_MC.csv")
loglogplot_composition <- as_vector(loglogplot_composition)
# Plot a scale free topology plot
WGCNA::scaleFreePlot(loglogplot_composition)

#Microbiome function
loglogplot_function <- read.csv("data_object_figS9_MC.csv")
loglogplot_function <- as_vector(loglogplot_function)
# Plot a scale free topology plot
WGCNA::scaleFreePlot(loglogplot_function)
```




###
ENDENDEND
###

























































#############################################################
# X.Integrated analysis2 - Ridge regression using three omics
#############################################################


#############################
#Main analysis (Outcome <- PPV use)
#############################
#data preparation
```{r}
library(caret); 
library(glmnet)
library(dplyr)
        
###################################################
MBmodule <- grep("^MB", names(df.all), value = TRUE)
Hostmodule <- grep("^Host", names(df.all), value = TRUE)
TXmodule <- grep("^TX", names(df.all), value = TRUE)
###################################################

#################Preparation#######################################
########## DEFINE EXPOSURE and REF LEVEL ##########################
df.all$CPAPintubate<- relevel(as.factor(df.all$CPAPintubate), "0") 
#df_pre$IntensiveTreatment<- relevel(as.factor(df_pre$IntensiveTreatment), "0") 
df_pre <- df.all %>% dplyr::select(c("CPAPintubate", 
                                     "Age_mo",
                                     "intake_sex",
                                     "RSVanyRV" ,  #  "RSVRVstatus_NPA"
                                     all_of(c(Hostmodule, TXmodule, MBmodule)))) #"CPAPintubate"

#df_pre$RSVRVstatus_NPA <- as.factor(df_pre$RSVRVstatus_NPA )
df_pre$RSVanyRV <- as.factor(df_pre$RSVanyRV)
df_pre$Y <- ifelse(df.all$CPAPintubate == 1, "Severe", "Nonsevere") %>% as.factor()
df_models <- df_pre %>% 
             dplyr::select(-c(CPAPintubate)) 
dim(df_models)
names(df_models)
###################################################################


#HF
names(df_pre)[which( names(df_pre) =="Hostred" ) ] <- "T cell regulation module"
names(df_pre)[which( names(df_pre)=="Hostblue" ) ] <- "Neutrophi/IL-1 module"
names(df_pre)[which( names(df_pre)=="Hostgreen" ) ] <- "GPCR module"
names(df_pre)[which( names(df_pre)=="Hostbrown" ) ] <- "Type I IFN module"
names(df_pre)[which( names(df_pre)=="Hostturquoise" ) ] <- "HF-1 module"
#MT
names(df_pre)[which( names(df_pre)=="TXdarkred" ) ] <- "S. pneumoniae/S. aureus module"
names(df_pre)[which( names(df_pre)=="TXdarkorchid" ) ] <- "MC-1 module"
names(df_pre)[which( names(df_pre)=="TXdarkturquoise" ) ] <- "Moraxella module"
names(df_pre)[which( names(df_pre)=="TXdarkblue" ) ] <- "Streptcoccus module"
names(df_pre)[which( names(df_pre)=="TXdarkorange" ) ] <- "Haemophilus module"
#MF
names(df_pre)[which( names(df_pre)=="MBplum2" ) ] <- "Plasma membrane module"
names(df_pre)[which( names(df_pre)=="MBlightcyan3" ) ] <- "mRNA metabolism module"
names(df_pre)[which( names(df_pre)=="MBroyalblue" ) ] <- "BCAA metabolism module"
names(df_pre)[which( names(df_pre)=="MBsalmon" ) ] <- "Oxidative stress response module"
names(df_pre)[which( names(df_pre)=="MBmediumpurple" ) ] <- "NADH dehydrogenase module"

#df_models_n <- df_models
#
#n <- length(df_models_n)
#for (i in 1:18) {
#  df_models_n[,i] <- as.numeric(df_models_n[,i])
#}

#df_models_s <- apply(df_models_n[,c(1:18)], 2, normalized)

#df_models_new <- cbind(df_models[,19], df_models_s)%>% as.data.frame()
#df_models <- df_models_new
#names(df_models)[ which( names(df_models )=="V1" ) ] <- "Y"
#df_models$Y <- df_models$Y %>% as.factor()
#df_models$Y <- ifelse(df_models$Y == 1, "NonSevere", "Severe") %>% as.factor()
```

#Ridge regression
```{r}
library(caret)
start.time<-proc.time() #開始
# Sanity check -- no NA alllowed
sum(is.na(df_models))
#Control the computational nuances of the train function
myTrainingControl <- trainControl(method  = "LOOCV", #"LOOCV", # "cv", "repeatedcv"
                                  #number  = 1,  #244
                                  #repeats = 5, #
                                  returnResamp="all",
                                  savePredictions = TRUE, 
                                  classProbs = TRUE, # Estimate class probabilities
                                  summaryFunction = twoClassSummary,
                                  #sampling="smote",  # subsampling for class imbalance
                                  preProcOptions = NULL)

#Ref:https://www.analyticsvidhya.com/blog/2016/12/practical-guide-to-implement-machine-learning-with-caret-package-in-r-with-practice-problem/#:~:text=8.-,Predictions%20using%20Caret,prob%E2%80%9D%20or%20%E2%80%9Craw%E2%80%9D.
# Make grid
train_grid.ridge <- expand.grid(alpha  = 0, lambda = seq(0.5, 0.6,  by=0.001) )

#Sanity check
class(df_models$intake_sex)
#df_models$Age_mo <- as.factor(df_models$Age_mo)
df_models$Y <- as.factor(df_models$Y)
df_models$intake_sex <-  as.factor(df_models$intake_sex)
df_models$RSVanyRV <-  as.factor(df_models$RSVanyRV)

table(df_models$Y, useNA = c("always"))
is.factor(df_models$Y)

set.seed(1128)
fit_ridge <- caret::train(Y ~ .,  
                              data   = df_models, 
                              method = "glmnet", #"gbm"
                              metric = "ROC",   #"deviance",
                              tuneGrid  = train_grid.ridge,
                              #preProcess= c("center", "scale" ), #"center", "scale" 
                              trControl = myTrainingControl
                               )

fit_ridge

############
# Decide model to evaluate
model_ridge_final <- fit_ridge 
###########
# Check the fitted model
plot(model_ridge_final)

#function get best result 
get_best_result2(model_ridge_final)
best_lambda <-  get_best_result2(model_ridge_final)[,2]
#best_lambda <- 0.20
# best parameter
model_ridge_final$bestTune
#Result of best parameter1
(ridge_full.or <- exp(coef(model_ridge_final$finalModel, s = best_lambda )))



#Result of best parameter2
mat_coeff_ridge <- coef(model_ridge_final$finalModel,
                              model_ridge_final$bestTune$lambda) 
mat_coeff_ridge <- mat_coeff_ridge[-1,] %>%   # Remove the intercept
                                        as.matrix()       # Still include batch indicator
sum(mat_coeff_ridge != 0)  # number of non-zero coeff.
## Check metabolites with non-zero coefficients (bring the metabolite names by mapping sequential ids)
df_coeff_ridge.pre <- mat_coeff_ridge %>% as.data.frame()
df_coeff_ridge <- rownames_to_column(df_coeff_ridge.pre, var = "variable")
names(df_coeff_ridge)[ which( names(df_coeff_ridge)=="V1") ] <- "coef"
df_coeff_ridge$OR <- exp(df_coeff_ridge$coef)
df_coeff_ridge

#1と２同じ結果です
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

#Confidence interval Bootstrap originial 95%CI percentile in three omics
```{r eval=FALSE}
start.time<-proc.time() #開始

#Preparation###########################
modulename <- df_coeff_ridge[,1]
nsim <- 2000

nofv <- length(modulename)
#Create Data frame
data_boot_holder <- createdf(nsim,nofv, colnames = c( modulename))
result_or <- createdf(nofv,7, colnames = c( "Module","OR","midCI","lowCI","upCI", "Coef", "absCoef"))
######################################

#loop
set.seed(1128)
data_boot <- df_models
for (i in 1:nsim)
{
   boot_rows <-sample(x=c(1:nrow(data_boot)), size=nrow(data_boot), replace=T)
   data_boot2 <-data_boot[boot_rows,]
   boot.ridge_full  <-   caret::train(Y ~ .,  
                              data   = data_boot2, 
                              method = "glmnet", #"gbm"
                              metric = "ROC",   #"deviance",
                              tuneGrid = expand.grid(lambda = best_lambda , alpha = 0),
                              #preProcess= c("center", "scale"),
                              trControl = myTrainingControl
                               )
   coef <- coef(boot.ridge_full$finalModel, s = best_lambda)
   #boot.ridge_full.or<- exp(coef(boot.ridge_full$finalModel, s = best_lambda ))

 for (ii in 1:nofv) {
   data_boot_holder[i,ii] <- coef[ii+1,1]
 }}
   
#Extract 
write.csv(data_boot_holder, "data_boot_holder_all_adjusted_r1.csv")

#Visualize distribution of coef
par(mfrow = c(3, 3))
for (i in 1:nofv) {
  hist(data_boot_holder[,i], col = 'skyblue3', breaks = 100)
}

# Extract Odds Ratio -Confidence Interval
for (i in 1:nofv) {
  result_or[i,1] <-  df_coeff_ridge$variable[i] 
  result_or[i,2] <-  df_coeff_ridge[i,3]
  d = quantile(data_boot_holder[,i], c(0.025,0.5, 0.975),na.rm=T)
  result_or[i,3] <- exp(d[2]) #0.5CI
  result_or[i,4] <- exp(d[1]) #0.025CI
  result_or[i,5] <- exp(d[3]) #0.975CI
  result_or[i,6] <- (d[2]) #0.5CI
  result_or[i,7] <- abs(d[2]) #0.5CI
}
result_or

#Extract 
write.csv(result_or, "result_ridge_all_adjusted_r1.csv")
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```
##############################
### Figure 5A ################
##############################
```{r}
df_result <- read.csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/Revision1_nc/result_ridge_all_adjusted_1004_2.cs")
#HF
df_result$Module[which( df_result$Module =="Hostred" ) ] <- "T cell regulation module"
df_result$Module[which( df_result$Module=="Hostblue" ) ] <- "Neutrophi/IL-1 module"
df_result$Module[which( df_result$Module=="Hostgreen" ) ] <- "GPCR module"
df_result$Module[which( df_result$Module=="Hostbrown" ) ] <- "Type I IFN module"
df_result$Module[which( df_result$Module=="Hostturquoise" ) ] <- "HF-1 module"
#MT
df_result$Module[which( df_result$Module=="TXdarkred" ) ] <- "S. pneumoniae/S. aureus module"
df_result$Module[which( df_result$Module=="TXdarkorchid" ) ] <- "MC-1 module"
df_result$Module[which( df_result$Module=="TXdarkturquoise" ) ] <- "Moraxella module"
df_result$Module[which( df_result$Module=="TXdarkblue" ) ] <- "Streptcoccus module"
df_result$Module[which( df_result$Module=="TXdarkorange" ) ] <- "Haemophilus module"
#MF
df_result$Module[which( df_result$Module=="MBplum2" ) ] <- "Plasma membrane module"
df_result$Module[which( df_result$Module=="MBlightcyan3" ) ] <- "mRNA metabolism module"
df_result$Module[which( df_result$Module=="MBroyalblue" ) ] <- "BCAA metabolism module"
df_result$Module[which( df_result$Module=="MBsalmon" ) ] <- "Oxidative stress response module"
df_result$Module[which( df_result$Module=="MBmediumpurple" ) ] <- "NADH dehydrogenase module"

#Lollipop chart
library(ggplot2)
df_result2 <- df_result[!(df_result$Module =="Age_mo"| df_result$Module =="intake_sex2"|df_result$Module =="RSVanyRV1"|df_result$Module =="RSVanyRV2"), ]


df_result2$specimen <- c("HF","HF","HF","HF","HF", "MT", "MT","MT", "MT", "MT", "MF","MF","MF","MF","MF")
df_result2$specimen  <- factor(df_result2$specimen, levels = c("HF", "MT", "MF"))

#df_coeff_ridge.cpap3$variable <- rownames(df_coeff_ridge.cpap3)
#df_result2$Module  <-  gsub("`","",result.adjust.rsv2$Module) 
#df_result2$Module  <-  gsub("Host","HF_", df_result2$Module ) 
#df_result2$Module  <-  gsub("TX","MT_", df_result2$Module ) 
#df_result2$Module  <-  gsub("MB","MF_", df_result2$Module ) 


#df_result2 <- df_result2[order(-df_result2$midCI ),] 
df_result3  <- data.frame(df_result2$Module, round(df_result2[,4:6],2), df_result2$specimen, df_result2$Coef, df_result2$absCoef)

names(df_result3)[which(names(df_result3)=="df_result2.Module") ] <- "Module"
names(df_result3)[which(names(df_result3)=="df_result2.specimen") ] <- "specimen"
names(df_result3)[which(names(df_result3)=="df_result2.Coef") ] <- "Coef"
names(df_result3)[which(names(df_result3)=="df_result2.absCoef") ] <- "absCoef"


df_result3$Module_f <- factor(df_result3$Module, levels=c("NADH dehydrogenase module" , "Oxidative stress response module"    ,"BCAA metabolism module" ,"mRNA metabolism module","Plasma membrane module","Haemophilus module" ,"Streptcoccus module" , "Moraxella module"  ,"MC-1 module"   ,"S. pneumoniae/S. aureus module","HF-1 module" , "Type I IFN module","GPCR module" ,"Neutrophi/IL-1 module"   ,"T cell regulation module" ))



df_result3 <- df_result3 %>% mutate(col= ifelse(df_result3$Coef < 0, "0", "1"))



# reorder is close to order, but is made to change the order of the factor levels.
#df_result3$Module = with(df_result3, reorder(Module, midCI))

df_result3 <- df_result3 %>% mutate(col= ifelse(df_result3$Coef < 0, "0", "1"))

df_result3$Module_f <- factor(df_result3$Module, levels=c("NADH dehydrogenase module" , "Oxidative stress response module"    ,"BCAA metabolism module" ,"mRNA metabolism module","Plasma membrane module","Haemophilus module" ,"Streptcoccus module" , "Moraxella module"  ,"MC-1 module"   ,"S. pneumoniae/S. aureus module","HF-1 module" , "Type I IFN module","GPCR module" ,"Neutrophi/IL-1 module"   ,"T cell regulation module" ))


P1 <- 
  df_result3 %>%
  ggplot(aes(Module_f,as.numeric(midCI))) +
  geom_segment( aes(x=Module_f, xend=Module_f, y=1, yend=as.numeric(midCI), color= col)) +
  geom_point( aes(color = col, size=as.numeric(absCoef)), fill=alpha("black", 1), shape=21, stroke=0) +
  theme_light()  +
  scale_color_manual(values=c("#4575B4", "#D73027"), 
                     labels = c("Negative", "Positive"), 
                     name = "Module") + #c("#F6EC5F", "#487AB9", "#DD4F40")
  #c("#ffce25", "#0500ad", "#f12e00")c("#E92600", "#FDF360", "#7BB7F8")
  scale_fill_manual(values=c("#4575B4", "#D73027")) +
  theme(legend.position = "bottom",
                                #panel.grid = element_blank(),
                                panel.grid.major.x  = element_line(colour = alpha("black", 0.22)),
                                panel.grid.minor.x  =element_blank(),
                                panel.grid.major.y   =element_blank(),
                                panel.grid.minor.y   =element_blank(),
                                panel.background = element_blank(),
                                text = element_text(size=9),
                                plot.caption = element_text(hjust = 0, face= "plain"),
                                axis.text.x = element_text(face= "plain", colour = "black", size=9),
                                axis.text.y = element_text(face= "plain", colour = "black", size=9, hjust=0),
                                axis.title.x = element_text(size=9),
                                axis.title.y = element_text(size=9),
                                legend.title = element_text(colour="black", size=9,  face="plain"),
                                legend.text = element_text(colour="black", size=9, face="plain"),
                                legend.background = element_rect(fill="grey90", colour = "white"),
                                panel.border = element_blank(),
                                axis.ticks.x = element_blank())+
  geom_hline(yintercept = 1, linetype = 1, color = "lightgray") +
  coord_flip() +
  labs(x=" ",y="Odds ratio for PPV use", title="")   # + guides(size = FALSE, color =F, alpha=F, col=F)


(P2<- P1    
            + scale_y_log10(limits = c(0.2,5.0), breaks = c(0.2,0.4,0.6,1.0,2.0,3.0,4.0,5.0)) 
            #+ annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 5.5, xmax = 10.5, fill = "black", alpha = 0.1) #, colour = "grey10"
            + annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 0.5, xmax = 5.5, fill = "grey80", alpha = 0.2)
            + annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 10.5, xmax = 15.7, fill = "grey80", alpha = 0.2)
  + geom_point( aes(color = col, size=as.numeric(absCoef), fill=col, alpha=0.8), shape=21, stroke=1)          
  + ggtitle("Adjusted analysis") 
            + theme(plot.title = element_text(hjust = 0.2))+ guides(size = FALSE, color =F, alpha=F, fill=F))


          
ppi=300
par(mar = c(8, 16, 2, 4)); # c(lower ,left, upper, right)
tiff(paste0 ("Ridge_adjusted_cpap_lollipop_r1.tiff"), height = 4*ppi, width = 6*ppi, res=ppi)
P2
dev.off()
```


##############################
#Sensitivity analysis 1 (Add race as confounder)
##############################
#data preparation
```{r}
library(caret); 
library(glmnet)
library(dplyr)
        
#################
MBmodule <- grep("^MB", names(df.all), value = TRUE)
Hostmodule <- grep("^Host", names(df.all), value = TRUE)
TXmodule <- grep("^TX", names(df.all), value = TRUE)
#################

#################Preparation#################
########## DEFINE EXPOSURE and REF LEVEL ##########################
df.all$CPAPintubate<- relevel(as.factor(df.all$CPAPintubate), "0") 
#df_pre$IntensiveTreatment<- relevel(as.factor(df_pre$IntensiveTreatment), "0") 
df_pre <- df.all %>% dplyr::select(c("CPAPintubate", 
                                     "Age_mo",
                                     "intake_sex",
                                     "RSVanyRV",
                                     "raceethn",  #  "RSVRVstatus_NPA"
                                     all_of(c(Hostmodule, TXmodule, MBmodule)))) #"CPAPintubate"

#df_pre$RSVRVstatus_NPA <- as.factor(df_pre$RSVRVstatus_NPA )
df_pre$RSVanyRV <- as.factor(df_pre$RSVanyRV)
df_pre$raceethn <- as.factor(df_pre$raceethn)
df_pre$Y <- ifelse(df.all$CPAPintubate == 1, "Severe", "Nonsevere") %>% as.factor()


df_models <- df_pre %>% 
             dplyr::select(-c(CPAPintubate)) 
dim(df_models)
names(df_models)
###################################################################

#df_models_n <- df_models
#
#n <- length(df_models_n)
#for (i in 1:18) {
#  df_models_n[,i] <- as.numeric(df_models_n[,i])
#}

#df_models_s <- apply(df_models_n[,c(1:18)], 2, normalized)

#df_models_new <- cbind(df_models[,19], df_models_s)%>% as.data.frame()
#df_models <- df_models_new
#names(df_models)[ which( names(df_models )=="V1" ) ] <- "Y"
#df_models$Y <- df_models$Y %>% as.factor()
#df_models$Y <- ifelse(df_models$Y == 1, "NonSevere", "Severe") %>% as.factor()
```

#Ridge regression
```{r}
library(caret)
start.time<-proc.time() #開始
# Sanity check -- no NA alllowed
sum(is.na(df_models))
#Control the computational nuances of the train function
myTrainingControl <- trainControl(method  = "LOOCV", #"LOOCV", # "cv", "repeatedcv"
                                  #number  = 1,  #244
                                  #repeats = 5, #
                                  returnResamp="all",
                                  savePredictions = TRUE, 
                                  classProbs = TRUE, # Estimate class probabilities
                                  summaryFunction = twoClassSummary,
                                  #sampling="smote",  # subsampling for class imbalance
                                  preProcOptions = NULL)

#Ref:https://www.analyticsvidhya.com/blog/2016/12/practical-guide-to-implement-machine-learning-with-caret-package-in-r-with-practice-problem/#:~:text=8.-,Predictions%20using%20Caret,prob%E2%80%9D%20or%20%E2%80%9Craw%E2%80%9D.
# Make grid
train_grid.ridge <- expand.grid(alpha  = 0, lambda = seq(0.5, 0.6,  by=0.001) )

#Sanity check
class(df_models$intake_sex)
#df_models$Age_mo <- as.factor(df_models$Age_mo)
df_models$Y <- as.factor(df_models$Y)
df_models$intake_sex <-  as.factor(df_models$intake_sex)
df_models$RSVanyRV <-  as.factor(df_models$RSVanyRV)

table(df_models$Y, useNA = c("always"))
is.factor(df_models$Y)

set.seed(1128)
fit_ridge <- caret::train(Y ~ .,  
                              data   = df_models, 
                              method = "glmnet", #"gbm"
                              metric = "ROC",   #"deviance",
                              tuneGrid  = train_grid.ridge,
                              #preProcess= c("center", "scale" ), #"center", "scale" 
                              trControl = myTrainingControl
                               )

fit_ridge

############
# Decide model to evaluate
model_ridge_final <- fit_ridge 
###########
# Check the fitted model
plot(model_ridge_final)

#function get best result 
get_best_result2(model_ridge_final)
best_lambda <-  get_best_result2(model_ridge_final)[,2]
#best_lambda <- 0.20
# best parameter
model_ridge_final$bestTune
#Result of best parameter1
(ridge_full.or <- exp(coef(model_ridge_final$finalModel, s = best_lambda )))



#Result of best parameter2
mat_coeff_ridge <- coef(model_ridge_final$finalModel,
                              model_ridge_final$bestTune$lambda) 
mat_coeff_ridge <- mat_coeff_ridge[-1,] %>%   # Remove the intercept
                                        as.matrix()       # Still include batch indicator
sum(mat_coeff_ridge != 0)  # number of non-zero coeff.
## Check metabolites with non-zero coefficients (bring the metabolite names by mapping sequential ids)
df_coeff_ridge.pre <- mat_coeff_ridge %>% as.data.frame()
df_coeff_ridge <- rownames_to_column(df_coeff_ridge.pre, var = "variable")
names(df_coeff_ridge)[ which( names(df_coeff_ridge)=="V1") ] <- "coef"
df_coeff_ridge$OR <- exp(df_coeff_ridge$coef)
df_coeff_ridge

#1と２同じ結果です
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

#Confidence interval Bootstrap originial 95%CI percentile in three omics
```{r eval=FALSE}
start.time<-proc.time() #開始

#Preparation###########################
modulename <- df_coeff_ridge[,1]
nsim <- 2000

nofv <- length(modulename)
#Create Data frame
data_boot_holder <- createdf(nsim,nofv, colnames = c( modulename))
result_or <- createdf(nofv,7, colnames = c( "Module","OR","midCI","lowCI","upCI", "Coef", "absCoef"))
######################################

#loop
set.seed(1128)
data_boot <- df_models
for (i in 1:nsim)
{
   boot_rows <-sample(x=c(1:nrow(data_boot)), size=nrow(data_boot), replace=T)
   data_boot2 <-data_boot[boot_rows,]
   boot.ridge_full  <-   caret::train(Y ~ .,  
                              data   = data_boot2, 
                              method = "glmnet", #"gbm"
                              metric = "ROC",   #"deviance",
                              tuneGrid = expand.grid(lambda = best_lambda , alpha = 0),
                              #preProcess= c("center", "scale"),
                              trControl = myTrainingControl
                               )
   coef <- coef(boot.ridge_full$finalModel, s = best_lambda)
   #boot.ridge_full.or<- exp(coef(boot.ridge_full$finalModel, s = best_lambda ))

 for (ii in 1:nofv) {
   data_boot_holder[i,ii] <- coef[ii+1,1]
 }}
   
#Extract 
write.csv(data_boot_holder, "data_boot_holder_all_adjusted_race_r1.csv")

#Visualize distribution of coef
par(mfrow = c(3, 3))
for (i in 1:nofv) {
  hist(data_boot_holder[,i], col = 'skyblue3', breaks = 100)
}

# Extract Odds Ratio -Confidence Interval
for (i in 1:nofv) {
  result_or[i,1] <-  df_coeff_ridge$variable[i] 
  result_or[i,2] <-  df_coeff_ridge[i,3]
  d = quantile(data_boot_holder[,i], c(0.025,0.5, 0.975),na.rm=T)
  result_or[i,3] <- exp(d[2]) #0.5CI
  result_or[i,4] <- exp(d[1]) #0.025CI
  result_or[i,5] <- exp(d[3]) #0.975CI
  result_or[i,6] <- (d[2]) #0.5CI
  result_or[i,7] <- abs(d[2]) #0.5CI
}
result_or

#Extract 
write.csv(result_or, "result_ridge_all_adjusted_race_r1.csv")
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

##############################
### Supplementary Figure 3 ###
##############################
```{r}
setwd("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/Revision1_nc")
df_result <- read.csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/Revision1_nc/result_ridge_all_adjusted_race_r1.csv")
#HF
df_result$Module[which( df_result$Module =="Hostred" ) ] <- "T cell regulation module"
df_result$Module[which( df_result$Module=="Hostblue" ) ] <- "Neutrophi/IL-1 module"
df_result$Module[which( df_result$Module=="Hostgreen" ) ] <- "GPCR module"
df_result$Module[which( df_result$Module=="Hostbrown" ) ] <- "Type I IFN module"
df_result$Module[which( df_result$Module=="Hostturquoise" ) ] <- "HF-1 module"
#MT
df_result$Module[which( df_result$Module=="TXdarkred" ) ] <- "S. pneumoniae/S. aureus module"
df_result$Module[which( df_result$Module=="TXdarkorchid" ) ] <- "MC-1 module"
df_result$Module[which( df_result$Module=="TXdarkturquoise" ) ] <- "Moraxella module"
df_result$Module[which( df_result$Module=="TXdarkblue" ) ] <- "Streptcoccus module"
df_result$Module[which( df_result$Module=="TXdarkorange" ) ] <- "Haemophilus module"
#MF
df_result$Module[which( df_result$Module=="MBplum2" ) ] <- "Plasma membrane module"
df_result$Module[which( df_result$Module=="MBlightcyan3" ) ] <- "mRNA metabolism module"
df_result$Module[which( df_result$Module=="MBroyalblue" ) ] <- "BCAA metabolism module"
df_result$Module[which( df_result$Module=="MBsalmon" ) ] <- "Oxidative stress response module"
df_result$Module[which( df_result$Module=="MBmediumpurple" ) ] <- "NADH dehydrogenase module"

#Lollipop chart
library(ggplot2)
df_result2 <- df_result[!(df_result$Module =="Age_mo"| df_result$Module =="intake_sex2"|df_result$Module =="RSVanyRV1"|df_result$Module =="RSVanyRV2"|df_result$Module =="raceethn2" |df_result$Module =="raceethn3" |df_result$Module =="raceethn4" ), ]


df_result2$specimen <- c("HF","HF","HF","HF","HF", "MT", "MT","MT", "MT", "MT", "MF","MF","MF","MF","MF")
df_result2$specimen  <- factor(df_result2$specimen, levels = c("HF", "MT", "MF"))

#df_coeff_ridge.cpap3$variable <- rownames(df_coeff_ridge.cpap3)
#df_result2$Module  <-  gsub("`","",result.adjust.rsv2$Module) 
#df_result2$Module  <-  gsub("Host","HF_", df_result2$Module ) 
#df_result2$Module  <-  gsub("TX","MT_", df_result2$Module ) 
#df_result2$Module  <-  gsub("MB","MF_", df_result2$Module ) 


#df_result2 <- df_result2[order(-df_result2$midCI ),] 
df_result3  <- data.frame(df_result2$Module, round(df_result2[,4:6],2), df_result2$specimen, df_result2$Coef, df_result2$absCoef)

names(df_result3)[which(names(df_result3)=="df_result2.Module") ] <- "Module"
names(df_result3)[which(names(df_result3)=="df_result2.specimen") ] <- "specimen"
names(df_result3)[which(names(df_result3)=="df_result2.Coef") ] <- "Coef"
names(df_result3)[which(names(df_result3)=="df_result2.absCoef") ] <- "absCoef"


df_result3$Module_f <- factor(df_result3$Module, levels=c("NADH dehydrogenase module" , "Oxidative stress response module"    ,"BCAA metabolism module" ,"mRNA metabolism module","Plasma membrane module","Haemophilus module" ,"Streptcoccus module" , "Moraxella module"  ,"MC-1 module"   ,"S. pneumoniae/S. aureus module","HF-1 module" , "Type I IFN module","GPCR module" ,"Neutrophi/IL-1 module"   ,"T cell regulation module" ))



df_result3 <- df_result3 %>% mutate(col= ifelse(df_result3$Coef < 0, "0", "1"))



# reorder is close to order, but is made to change the order of the factor levels.
#df_result3$Module = with(df_result3, reorder(Module, midCI))

df_result3 <- df_result3 %>% mutate(col= ifelse(df_result3$Coef < 0, "0", "1"))

df_result3$Module_f <- factor(df_result3$Module, levels=c("NADH dehydrogenase module" , "Oxidative stress response module"    ,"BCAA metabolism module" ,"mRNA metabolism module","Plasma membrane module","Haemophilus module" ,"Streptcoccus module" , "Moraxella module"  ,"MC-1 module"   ,"S. pneumoniae/S. aureus module","HF-1 module" , "Type I IFN module","GPCR module" ,"Neutrophi/IL-1 module"   ,"T cell regulation module" ))


P1 <- 
  df_result3 %>%
  ggplot(aes(Module_f,as.numeric(midCI))) +
  geom_segment( aes(x=Module_f, xend=Module_f, y=1, yend=as.numeric(midCI), color= col)) +
  geom_point( aes(color = col, size=as.numeric(absCoef)), fill=alpha("black", 1), shape=21, stroke=0) +
  theme_light()  +
  scale_color_manual(values=c("#4575B4", "#D73027"), 
                     labels = c("Negative", "Positive"), 
                     name = "Module") + #c("#F6EC5F", "#487AB9", "#DD4F40")
  #c("#ffce25", "#0500ad", "#f12e00")c("#E92600", "#FDF360", "#7BB7F8")
  scale_fill_manual(values=c("#4575B4", "#D73027")) +
  theme(legend.position = "bottom",
                                #panel.grid = element_blank(),
                                panel.grid.major.x  = element_line(colour = alpha("black", 0.22)),
                                panel.grid.minor.x  =element_blank(),
                                panel.grid.major.y   =element_blank(),
                                panel.grid.minor.y   =element_blank(),
                                panel.background = element_blank(),
                                text = element_text(size=9),
                                plot.caption = element_text(hjust = 0, face= "plain"),
                                axis.text.x = element_text(face= "plain", colour = "black", size=9),
                                axis.text.y = element_text(face= "plain", colour = "black", size=9, hjust=0),
                                axis.title.x = element_text(size=9),
                                axis.title.y = element_text(size=9),
                                legend.title = element_text(colour="black", size=9,  face="plain"),
                                legend.text = element_text(colour="black", size=9, face="plain"),
                                legend.background = element_rect(fill="grey90", colour = "white"),
                                panel.border = element_blank(),
                                axis.ticks.x = element_blank())+
  geom_hline(yintercept = 1, linetype = 1, color = "lightgray") +
  coord_flip() +
  labs(x=" ",y="Odds ratio for PPV use", title="")   # + guides(size = FALSE, color =F, alpha=F, col=F)


(P2<- P1    
            + scale_y_log10(limits = c(0.2,5.0), breaks = c(0.2,0.4,0.6,1.0,2.0,3.0,4.0,5.0)) 
            #+ annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 5.5, xmax = 10.5, fill = "black", alpha = 0.1) #, colour = "grey10"
            + annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 0.5, xmax = 5.5, fill = "grey80", alpha = 0.2)
            + annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 10.5, xmax = 15.7, fill = "grey80", alpha = 0.2)
  + geom_point( aes(color = col, size=as.numeric(absCoef), fill=col, alpha=0.8), shape=21, stroke=1)          
  + ggtitle("Adjusted analysis") 
            + theme(plot.title = element_text(hjust = 0.2))+ guides(size = FALSE, color =F, alpha=F, fill=F))


          
ppi=300
par(mar = c(8, 16, 2, 4)); # c(lower ,left, upper, right)
tiff(paste0 ("Ridge_adjusted_cpap_lollipop_race_r1.tiff"), height = 4*ppi, width = 6*ppi, res=ppi)
P2
dev.off()
```


##############################
#Sensitivity analysis 2 (outcome = intensive treatment use)
##############################
#data preparation
```{r}
library(caret); 
library(glmnet)
library(dplyr)
        
#################
MBmodule <- grep("^MB", names(df.all), value = TRUE)
Hostmodule <- grep("^Host", names(df.all), value = TRUE)
TXmodule <- grep("^TX", names(df.all), value = TRUE)
#################

#################Preparation#################
########## DEFINE EXPOSURE and REF LEVEL ##########################
df.all$IntensiveTreatment<- relevel(as.factor(df.all$IntensiveTreatment), "0") 
df_pre <- df.all %>% dplyr::select(c("IntensiveTreatment", 
                                     "Age_mo",
                                     "intake_sex",
                                     "RSVanyRV" ,  #  "RSVRVstatus_NPA"
                                     all_of(c(Hostmodule, TXmodule, MBmodule)))) 

#df_pre$RSVRVstatus_NPA <- as.factor(df_pre$RSVRVstatus_NPA )
df_pre$RSVanyRV <- as.factor(df_pre$RSVanyRV)
df_pre$Y <- ifelse(df.all$IntensiveTreatment == 1, "Severe", "Nonsevere") %>% as.factor()
df_models <- df_pre %>% 
             dplyr::select(-c(IntensiveTreatment)) 
dim(df_models)
names(df_models)
###################################################################

#df_models_n <- df_models
#
#n <- length(df_models_n)
#for (i in 1:18) {
#  df_models_n[,i] <- as.numeric(df_models_n[,i])
#}

#df_models_s <- apply(df_models_n[,c(1:18)], 2, normalized)

#df_models_new <- cbind(df_models[,19], df_models_s)%>% as.data.frame()
#df_models <- df_models_new
#names(df_models)[ which( names(df_models )=="V1" ) ] <- "Y"
#df_models$Y <- df_models$Y %>% as.factor()
#df_models$Y <- ifelse(df_models$Y == 1, "NonSevere", "Severe") %>% as.factor()
```

#Ridge regression
```{r}
library(caret)
start.time<-proc.time() #開始
# Sanity check -- no NA alllowed
sum(is.na(df_models))
#Control the computational nuances of the train function
myTrainingControl <- trainControl(method  = "LOOCV", #"LOOCV", # "cv", "repeatedcv"
                                  #number  = 1,  #244
                                  #repeats = 5, #
                                  returnResamp="all",
                                  savePredictions = TRUE, 
                                  classProbs = TRUE, # Estimate class probabilities
                                  summaryFunction = twoClassSummary,
                                  #sampling="smote",  # subsampling for class imbalance
                                  preProcOptions = NULL)

#Ref:https://www.analyticsvidhya.com/blog/2016/12/practical-guide-to-implement-machine-learning-with-caret-package-in-r-with-practice-problem/#:~:text=8.-,Predictions%20using%20Caret,prob%E2%80%9D%20or%20%E2%80%9Craw%E2%80%9D.
# Make grid
train_grid.ridge <- expand.grid(alpha  = 0, lambda = seq(0.5, 0.6,  by=0.001) )

#Sanity check
class(df_models$intake_sex)
#df_models$Age_mo <- as.factor(df_models$Age_mo)
df_models$Y <- as.factor(df_models$Y)
df_models$intake_sex <-  as.factor(df_models$intake_sex)
df_models$RSVanyRV <-  as.factor(df_models$RSVanyRV)

table(df_models$Y, useNA = c("always"))
is.factor(df_models$Y)

set.seed(1128)
fit_ridge <- caret::train(Y ~ .,  
                              data   = df_models, 
                              method = "glmnet", #"gbm"
                              metric = "ROC",   #"deviance",
                              tuneGrid  = train_grid.ridge,
                              #preProcess= c("center", "scale" ), #"center", "scale" 
                              trControl = myTrainingControl
                               )

fit_ridge

############
# Decide model to evaluate
model_ridge_final <- fit_ridge 
###########
# Check the fitted model
plot(model_ridge_final)

#function get best result 
get_best_result2(model_ridge_final)
best_lambda <-  get_best_result2(model_ridge_final)[,2]
#best_lambda <- 0.20
# best parameter
model_ridge_final$bestTune
#Result of best parameter1
(ridge_full.or <- exp(coef(model_ridge_final$finalModel, s = best_lambda )))



#Result of best parameter2
mat_coeff_ridge <- coef(model_ridge_final$finalModel,
                              model_ridge_final$bestTune$lambda) 
mat_coeff_ridge <- mat_coeff_ridge[-1,] %>%   # Remove the intercept
                                        as.matrix()       # Still include batch indicator
sum(mat_coeff_ridge != 0)  # number of non-zero coeff.
## Check metabolites with non-zero coefficients (bring the metabolite names by mapping sequential ids)
df_coeff_ridge.pre <- mat_coeff_ridge %>% as.data.frame()
df_coeff_ridge <- rownames_to_column(df_coeff_ridge.pre, var = "variable")
names(df_coeff_ridge)[ which( names(df_coeff_ridge)=="V1") ] <- "coef"
df_coeff_ridge$OR <- exp(df_coeff_ridge$coef)
df_coeff_ridge

#1と２同じ結果です
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

#Confidence interval Bootstrap originial 95%CI percentile in three omics
```{r eval=FALSE}
start.time<-proc.time() #開始

#Preparation###########################
modulename <- df_coeff_ridge[,1]
nsim <- 2000

nofv <- length(modulename)
#Create Data frame
data_boot_holder <- createdf(nsim,nofv, colnames = c( modulename))
result_or <- createdf(nofv,7, colnames = c( "Module","OR","midCI","lowCI","upCI", "Coef", "absCoef"))
######################################

#loop
set.seed(1128)
data_boot <- df_models
for (i in 1:nsim)
{
   boot_rows <-sample(x=c(1:nrow(data_boot)), size=nrow(data_boot), replace=T)
   data_boot2 <-data_boot[boot_rows,]
   boot.ridge_full  <-   caret::train(Y ~ .,  
                              data   = data_boot2, 
                              method = "glmnet", #"gbm"
                              metric = "ROC",   #"deviance",
                              tuneGrid = expand.grid(lambda = best_lambda , alpha = 0),
                              #preProcess= c("center", "scale"),
                              trControl = myTrainingControl
                               )
   coef <- coef(boot.ridge_full$finalModel, s = best_lambda)
   #boot.ridge_full.or<- exp(coef(boot.ridge_full$finalModel, s = best_lambda ))

 for (ii in 1:nofv) {
   data_boot_holder[i,ii] <- coef[ii+1,1]
 }}
   
#Extract 
write.csv(data_boot_holder, "data_boot_holder_all_adjusted_icuuse_r1.csv")

#Visualize distribution of coef
par(mfrow = c(3, 3))
for (i in 1:nofv) {
  hist(data_boot_holder[,i], col = 'skyblue3', breaks = 100)
}

# Extract Odds Ratio -Confidence Interval
for (i in 1:nofv) {
  result_or[i,1] <-  df_coeff_ridge$variable[i] 
  result_or[i,2] <-  df_coeff_ridge[i,3]
  d = quantile(data_boot_holder[,i], c(0.025,0.5, 0.975),na.rm=T)
  result_or[i,3] <- exp(d[2]) #0.5CI
  result_or[i,4] <- exp(d[1]) #0.025CI
  result_or[i,5] <- exp(d[3]) #0.975CI
  result_or[i,6] <- (d[2]) #0.5CI
  result_or[i,7] <- abs(d[2]) #0.5CI
}
result_or

#Extract 
write.csv(result_or, "result_ridge_all_adjusted_icuuse_r1.csv")
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

##############################
### Supplementary Figure 4 ###
##############################

```{r}
df_result <- read.csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/finaldata/result_ridge_all_adjusted_icuuse_1004_2.csv")
#HF
df_result$Module[which( df_result$Module =="Hostred" ) ] <- "T cell regulation module"
df_result$Module[which( df_result$Module=="Hostblue" ) ] <- "Neutrophi/IL-1 module"
df_result$Module[which( df_result$Module=="Hostgreen" ) ] <- "GPCR module"
df_result$Module[which( df_result$Module=="Hostbrown" ) ] <- "Type I IFN module"
df_result$Module[which( df_result$Module=="Hostturquoise" ) ] <- "HF-1 module"
#MT
df_result$Module[which( df_result$Module=="TXdarkred" ) ] <- "S. pneumoniae/S. aureus module"
df_result$Module[which( df_result$Module=="TXdarkorchid" ) ] <- "MC-1 module"
df_result$Module[which( df_result$Module=="TXdarkturquoise" ) ] <- "Moraxella module"
df_result$Module[which( df_result$Module=="TXdarkblue" ) ] <- "Streptcoccus module"
df_result$Module[which( df_result$Module=="TXdarkorange" ) ] <- "Haemophilus module"
#MF
df_result$Module[which( df_result$Module=="MBplum2" ) ] <- "Plasma membrane module"
df_result$Module[which( df_result$Module=="MBlightcyan3" ) ] <- "mRNA metabolism module"
df_result$Module[which( df_result$Module=="MBroyalblue" ) ] <- "BCAA metabolism module"
df_result$Module[which( df_result$Module=="MBsalmon" ) ] <- "Oxidative stress response module"
df_result$Module[which( df_result$Module=="MBmediumpurple" ) ] <- "NADH dehydrogenase module"

## lollipop
library(ggplot2)
df_result2 <- df_result[!(df_result$Module =="Age_mo"| df_result$Module =="intake_sex2"|df_result$Module =="RSVanyRV1"|df_result$Module =="RSVanyRV2"), ]


df_result2$specimen <- c("HF","HF","HF","HF","HF", "MT", "MT","MT", "MT", "MT", "MF","MF","MF","MF","MF")
df_result2$specimen  <- factor(df_result2$specimen, levels = c("HF", "MT", "MF"))

#df_coeff_ridge.cpap3$variable <- rownames(df_coeff_ridge.cpap3)
#df_result2$Module  <-  gsub("`","",result.adjust.rsv2$Module) 
#df_result2$Module  <-  gsub("Host","HF_", df_result2$Module ) 
#df_result2$Module  <-  gsub("TX","MT_", df_result2$Module ) 
#df_result2$Module  <-  gsub("MB","MF_", df_result2$Module ) 


#df_result2 <- df_result2[order(-df_result2$midCI ),] 
df_result3  <- data.frame(df_result2$Module, round(df_result2[,4:6],2), df_result2$specimen, df_result2$Coef, df_result2$absCoef)

names(df_result3)[which(names(df_result3)=="df_result2.Module") ] <- "Module"
names(df_result3)[which(names(df_result3)=="df_result2.specimen") ] <- "specimen"
names(df_result3)[which(names(df_result3)=="df_result2.Coef") ] <- "Coef"
names(df_result3)[which(names(df_result3)=="df_result2.absCoef") ] <- "absCoef"

df_result3$Module_f <- factor(df_result3$Module, levels=c("NADH dehydrogenase module" , "Oxidative stress response module"    ,"BCAA metabolism module" ,"mRNA metabolism module","Plasma membrane module","Haemophilus module" ,"Streptcoccus module" , "Moraxella module"  ,"MC-1 module"   ,"S. pneumoniae/S. aureus module","HF-1 module" , "Type I IFN module","GPCR module" ,"Neutrophi/IL-1 module" ,"T cell regulation module" ))


df_result3 <- df_result3 %>% mutate(col= ifelse(df_result3$Coef < 0, "0", "1"))


P1 <- 
  df_result3 %>%
  ggplot(aes(Module_f, as.numeric(midCI))) +
  geom_segment( aes(x=Module_f, xend=Module_f, y=1, yend=as.numeric(midCI), color= col)) +
  geom_point( aes(color = col, size=as.numeric(absCoef)), fill=alpha("black", 1), shape=21, stroke=0) +
  theme_light()  +
  scale_color_manual(values=c("#4575B4", "#D73027"), 
                     labels = c("Negative", "Positive"), 
                     name = "Module") + #c("#F6EC5F", "#487AB9", "#DD4F40")
  #c("#ffce25", "#0500ad", "#f12e00")c("#E92600", "#FDF360", "#7BB7F8")
  scale_fill_manual(values=c("#4575B4", "#D73027")) +
  theme(legend.position = "bottom",
                                #panel.grid = element_blank(),
                                panel.grid.major.x  = element_line(colour = alpha("black", 0.22)),
                                panel.grid.minor.x  =element_blank(),
                                panel.grid.major.y   =element_blank(),
                                panel.grid.minor.y   =element_blank(),
                                panel.background = element_blank(),
                                text = element_text(size=9),
                                plot.caption = element_text(hjust = 0, face= "plain"),
                                axis.text.x = element_text(face= "plain", colour = "black", size=9),
                                axis.text.y = element_text(face= "plain", colour = "black", size=9, hjust=0),
                                axis.title.x = element_text(size=9),
                                axis.title.y = element_text(size=9),
                                legend.title = element_text(colour="black", size=9,  face="plain"),
                                legend.text = element_text(colour="black", size=9, face="plain"),
                                legend.background = element_rect(fill="grey90", colour = "white"),
                                panel.border = element_blank(),
                                axis.ticks.x = element_blank())+
  geom_hline(yintercept = 1, linetype = 1, color = "lightgray") +
  coord_flip() +
  labs(x=" ",y="Odds ratio for intensive care use", title="")   # + guides(size = FALSE, color =F, alpha=F, col=F)


(P2<- P1    
            + scale_y_log10(limits = c(0.2,8.0), breaks = c(0.2,0.4,0.6,1.0,2.0,4.0,8.0)) 
            #+ annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 5.5, xmax = 10.5, fill = "black", alpha = 0.1) #, colour = "grey10"
            + annotate(geom = "rect", ymin = 0.2, ymax = 8, xmin = 0.5, xmax = 5.5, fill = "grey80", alpha = 0.2)
            + annotate(geom = "rect", ymin = 0.2, ymax = 8, xmin = 10.5, xmax = 15.7, fill = "grey80", alpha = 0.2)
  + geom_point( aes(color = col, size=as.numeric(absCoef), fill=col, alpha=0.8), shape=21, stroke=1)          
  + ggtitle("Adjusted analysis") 
            + theme(plot.title = element_text(hjust = 0.2))+ guides(size = "none", color ="none", alpha="none", fill="none"))

ppi=300
par(mar = c(1, 1, 1, 1)); 
tiff(paste0 ("Ridge_adjusted_icuuse_lollipop_r1.tiff"), height = 4*ppi, width = 6*ppi, res=ppi)
P2
dev.off()
```



##############################
#Sensitivity analysis3 (outcome = intensive treatment use + Add race as confounder)
##############################

#data preparation
```{r}
library(caret); 
library(glmnet)
library(dplyr)
        
#################
MBmodule <- grep("^MB", names(df.all), value = TRUE)
Hostmodule <- grep("^Host", names(df.all), value = TRUE)
TXmodule <- grep("^TX", names(df.all), value = TRUE)
#################

#################Preparation#################
########## DEFINE EXPOSURE and REF LEVEL ##########################
df.all$IntensiveTreatment<- relevel(as.factor(df.all$IntensiveTreatment), "0") 
df_pre <- df.all %>% dplyr::select(c("IntensiveTreatment", 
                                     "Age_mo",
                                     "intake_sex",
                                     "RSVanyRV",  #  "RSVRVstatus_NPA"
                                     "raceethn",
                                     all_of(c(Hostmodule, TXmodule, MBmodule)))) 

#df_pre$RSVRVstatus_NPA <- as.factor(df_pre$RSVRVstatus_NPA )
df_pre$RSVanyRV <- as.factor(df_pre$RSVanyRV)
df_pre$raceethn <- as.factor(df_pre$raceethn)
df_pre$Y <- ifelse(df.all$IntensiveTreatment == 1, "Severe", "Nonsevere") %>% as.factor()
df_models <- df_pre %>% 
             dplyr::select(-c(IntensiveTreatment)) 
dim(df_models)
names(df_models)
###################################################################
```

#Ridge regression
```{r}
library(caret)
start.time<-proc.time() #開始
# Sanity check -- no NA alllowed
sum(is.na(df_models))
#Control the computational nuances of the train function
myTrainingControl <- trainControl(method  = "LOOCV", #"LOOCV", # "cv", "repeatedcv"
                                  #number  = 1,  #244
                                  #repeats = 5, #
                                  returnResamp="all",
                                  savePredictions = TRUE, 
                                  classProbs = TRUE, # Estimate class probabilities
                                  summaryFunction = twoClassSummary,
                                  #sampling="smote",  # subsampling for class imbalance
                                  preProcOptions = NULL)

#Ref:https://www.analyticsvidhya.com/blog/2016/12/practical-guide-to-implement-machine-learning-with-caret-package-in-r-with-practice-problem/#:~:text=8.-,Predictions%20using%20Caret,prob%E2%80%9D%20or%20%E2%80%9Craw%E2%80%9D.
# Make grid
train_grid.ridge <- expand.grid(alpha  = 0, lambda = seq(0.5, 0.6,  by=0.001) )

#Sanity check
class(df_models$intake_sex)
#df_models$Age_mo <- as.factor(df_models$Age_mo)
df_models$Y <- as.factor(df_models$Y)
df_models$intake_sex <-  as.factor(df_models$intake_sex)
df_models$RSVanyRV <-  as.factor(df_models$RSVanyRV)

table(df_models$Y, useNA = c("always"))
is.factor(df_models$Y)

set.seed(1128)
fit_ridge <- caret::train(Y ~ .,  
                              data   = df_models, 
                              method = "glmnet", #"gbm"
                              metric = "ROC",   #"deviance",
                              tuneGrid  = train_grid.ridge,
                              #preProcess= c("center", "scale" ), #"center", "scale" 
                              trControl = myTrainingControl
                               )

fit_ridge

############
# Decide model to evaluate
model_ridge_final <- fit_ridge 
###########
# Check the fitted model
plot(model_ridge_final)

#function get best result 
get_best_result2(model_ridge_final)
best_lambda <-  get_best_result2(model_ridge_final)[,2]
#best_lambda <- 0.20
# best parameter
model_ridge_final$bestTune
#Result of best parameter1
(ridge_full.or <- exp(coef(model_ridge_final$finalModel, s = best_lambda )))



#Result of best parameter2
mat_coeff_ridge <- coef(model_ridge_final$finalModel,
                              model_ridge_final$bestTune$lambda) 
mat_coeff_ridge <- mat_coeff_ridge[-1,] %>%   # Remove the intercept
                                        as.matrix()       # Still include batch indicator
sum(mat_coeff_ridge != 0)  # number of non-zero coeff.
## Check metabolites with non-zero coefficients (bring the metabolite names by mapping sequential ids)
df_coeff_ridge.pre <- mat_coeff_ridge %>% as.data.frame()
df_coeff_ridge <- rownames_to_column(df_coeff_ridge.pre, var = "variable")
names(df_coeff_ridge)[ which( names(df_coeff_ridge)=="V1") ] <- "coef"
df_coeff_ridge$OR <- exp(df_coeff_ridge$coef)
df_coeff_ridge

#1と２同じ結果です
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

#Confidence interval Bootstrap originial 95%CI percentile in three omics
```{r eval=FALSE}
start.time<-proc.time() #開始

#Preparation###########################
modulename <- df_coeff_ridge[,1]
nsim <- 2000

nofv <- length(modulename)
#Create Data frame
data_boot_holder <- createdf(nsim,nofv, colnames = c( modulename))
result_or <- createdf(nofv,7, colnames = c( "Module","OR","midCI","lowCI","upCI", "Coef", "absCoef"))
######################################

#loop
set.seed(1128)
data_boot <- df_models
for (i in 1:nsim)
{
   boot_rows <-sample(x=c(1:nrow(data_boot)), size=nrow(data_boot), replace=T)
   data_boot2 <-data_boot[boot_rows,]
   boot.ridge_full  <-   caret::train(Y ~ .,  
                              data   = data_boot2, 
                              method = "glmnet", #"gbm"
                              metric = "ROC",   #"deviance",
                              tuneGrid = expand.grid(lambda = best_lambda , alpha = 0),
                              #preProcess= c("center", "scale"),
                              trControl = myTrainingControl
                               )
   coef <- coef(boot.ridge_full$finalModel, s = best_lambda)
   #boot.ridge_full.or<- exp(coef(boot.ridge_full$finalModel, s = best_lambda ))

 for (ii in 1:nofv) {
   data_boot_holder[i,ii] <- coef[ii+1,1]
 }}
   
#Extract 
write.csv(data_boot_holder, "data_boot_holder_all_adjusted_icuuse_race_r1.csv")

#Visualize distribution of coef
par(mfrow = c(3, 3))
for (i in 1:nofv) {
  hist(data_boot_holder[,i], col = 'skyblue3', breaks = 100)
}

# Extract Odds Ratio -Confidence Interval
for (i in 1:nofv) {
  result_or[i,1] <-  df_coeff_ridge$variable[i] 
  result_or[i,2] <-  df_coeff_ridge[i,3]
  d = quantile(data_boot_holder[,i], c(0.025,0.5, 0.975),na.rm=T)
  result_or[i,3] <- exp(d[2]) #0.5CI
  result_or[i,4] <- exp(d[1]) #0.025CI
  result_or[i,5] <- exp(d[3]) #0.975CI
  result_or[i,6] <- (d[2]) #0.5CI
  result_or[i,7] <- abs(d[2]) #0.5CI
}
result_or

#Extract 
write.csv(result_or, "result_ridge_all_adjusted_icuuse_race_r1.csv")
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

##############################
### Supplementary Figure 5 ###
##############################
```{r}
setwd("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/Revision1_nc")
df_result <- read.csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/Revision1_nc/result_ridge_all_adjusted_icuuse_race_r1.csv")
#HF
df_result$Module[which( df_result$Module =="Hostred" ) ] <- "T cell regulation module"
df_result$Module[which( df_result$Module=="Hostblue" ) ] <- "Neutrophi/IL-1 module"
df_result$Module[which( df_result$Module=="Hostgreen" ) ] <- "GPCR module"
df_result$Module[which( df_result$Module=="Hostbrown" ) ] <- "Type I IFN module"
df_result$Module[which( df_result$Module=="Hostturquoise" ) ] <- "HF-1 module"
#MT
df_result$Module[which( df_result$Module=="TXdarkred" ) ] <- "S. pneumoniae/S. aureus module"
df_result$Module[which( df_result$Module=="TXdarkorchid" ) ] <- "MC-1 module"
df_result$Module[which( df_result$Module=="TXdarkturquoise" ) ] <- "Moraxella module"
df_result$Module[which( df_result$Module=="TXdarkblue" ) ] <- "Streptcoccus module"
df_result$Module[which( df_result$Module=="TXdarkorange" ) ] <- "Haemophilus module"
#MF
df_result$Module[which( df_result$Module=="MBplum2" ) ] <- "Plasma membrane module"
df_result$Module[which( df_result$Module=="MBlightcyan3" ) ] <- "mRNA metabolism module"
df_result$Module[which( df_result$Module=="MBroyalblue" ) ] <- "BCAA metabolism module"
df_result$Module[which( df_result$Module=="MBsalmon" ) ] <- "Oxidative stress response module"
df_result$Module[which( df_result$Module=="MBmediumpurple" ) ] <- "NADH dehydrogenase module"

#Lollipop chart
library(ggplot2)
df_result2 <- df_result[!(df_result$Module =="Age_mo"| df_result$Module =="intake_sex2"|df_result$Module =="RSVanyRV1"|df_result$Module =="RSVanyRV2"|df_result$Module =="raceethn2" |df_result$Module =="raceethn3" |df_result$Module =="raceethn4" ), ]


df_result2$specimen <- c("HF","HF","HF","HF","HF", "MT", "MT","MT", "MT", "MT", "MF","MF","MF","MF","MF")
df_result2$specimen  <- factor(df_result2$specimen, levels = c("HF", "MT", "MF"))

df_result3  <- data.frame(df_result2$Module, round(df_result2[,4:6],2), df_result2$specimen, df_result2$Coef, df_result2$absCoef)

names(df_result3)[which(names(df_result3)=="df_result2.Module") ] <- "Module"
names(df_result3)[which(names(df_result3)=="df_result2.specimen") ] <- "specimen"
names(df_result3)[which(names(df_result3)=="df_result2.Coef") ] <- "Coef"
names(df_result3)[which(names(df_result3)=="df_result2.absCoef") ] <- "absCoef"


df_result3$Module_f <- factor(df_result3$Module, levels=c("NADH dehydrogenase module" , "Oxidative stress response module"    ,"BCAA metabolism module" ,"mRNA metabolism module","Plasma membrane module","Haemophilus module" ,"Streptcoccus module" , "Moraxella module"  ,"MC-1 module"   ,"S. pneumoniae/S. aureus module","HF-1 module" , "Type I IFN module","GPCR module" ,"Neutrophi/IL-1 module"   ,"T cell regulation module" ))

df_result3 <- df_result3 %>% mutate(col= ifelse(df_result3$Coef < 0, "0", "1"))


P1 <- 
  df_result3 %>%
  ggplot(aes(Module_f,as.numeric(midCI))) +
  geom_segment( aes(x=Module_f, xend=Module_f, y=1, yend=as.numeric(midCI), color= col)) +
  geom_point( aes(color = col, size=as.numeric(absCoef)), fill=alpha("black", 1), shape=21, stroke=0) +
  theme_light()  +
  scale_color_manual(values=c("#4575B4", "#D73027"), 
                     labels = c("Negative", "Positive"), 
                     name = "Module") + #c("#F6EC5F", "#487AB9", "#DD4F40")
  #c("#ffce25", "#0500ad", "#f12e00")c("#E92600", "#FDF360", "#7BB7F8")
  scale_fill_manual(values=c("#4575B4", "#D73027")) +
  theme(legend.position = "bottom",
                                #panel.grid = element_blank(),
                                panel.grid.major.x  = element_line(colour = alpha("black", 0.22)),
                                panel.grid.minor.x  =element_blank(),
                                panel.grid.major.y   =element_blank(),
                                panel.grid.minor.y   =element_blank(),
                                panel.background = element_blank(),
                                text = element_text(size=9),
                                plot.caption = element_text(hjust = 0, face= "plain"),
                                axis.text.x = element_text(face= "plain", colour = "black", size=9),
                                axis.text.y = element_text(face= "plain", colour = "black", size=9, hjust=0),
                                axis.title.x = element_text(size=9),
                                axis.title.y = element_text(size=9),
                                legend.title = element_text(colour="black", size=9,  face="plain"),
                                legend.text = element_text(colour="black", size=9, face="plain"),
                                legend.background = element_rect(fill="grey90", colour = "white"),
                                panel.border = element_blank(),
                                axis.ticks.x = element_blank())+
  geom_hline(yintercept = 1, linetype = 1, color = "lightgray") +
  coord_flip() +
  labs(x=" ",y="Odds ratio for intensive care use", title="")   # + guides(size = FALSE, color =F, alpha=F, col=F)



(P2<- P1    
            + scale_y_log10(limits = c(0.2,8.0), breaks = c(0.2,0.4,0.6,1.0,2.0,4.0,8.0)) 
            #+ annotate(geom = "rect", ymin = 0.2, ymax = 5, xmin = 5.5, xmax = 10.5, fill = "black", alpha = 0.1) #, colour = "grey10"
            + annotate(geom = "rect", ymin = 0.2, ymax = 8, xmin = 0.5, xmax = 5.5, fill = "grey80", alpha = 0.2)
            + annotate(geom = "rect", ymin = 0.2, ymax = 8, xmin = 10.5, xmax = 15.7, fill = "grey80", alpha = 0.2)
  + geom_point( aes(color = col, size=as.numeric(absCoef), fill=col, alpha=0.8), shape=21, stroke=1)          
  + ggtitle("Adjusted analysis") 
            + theme(plot.title = element_text(hjust = 0.2))+ guides(size = "none", color ="none", alpha="none", fill="none"))
          
ppi=300
par(mar = c(8, 16, 2, 4)); 
tiff(paste0 ("Ridge_adjusted_icuuse_lollipop_race_r1.tiff"), height = 4*ppi, width = 6*ppi, res=ppi)
P2
dev.off()
```


##########################
#Host
#GO analysis in each module
##########################
#Figures
# axis label name color https://stackoverflow.com/questions/38862303/customize-ggplot2-axis-labels-with-different-colors
```{r}
library(scales) 
colors1=c( "#C7E9B4" ,"#C7E9B4" ,"#7FCDBB" ,"#41B6C4" ,"#1D91C0", "#225EA8", "#0C2C84")
color1 <- colorRampPalette(colors = colors1)(300)
colors2=c( " white", "blue")
color2 <- colorRampPalette(colors = colors2)(10)

library("scales")
reverselog_trans <- function(base = exp(1)) {
    trans <- function(x) -log(x, base)
    inv <- function(x) base^(-x)
    trans_new(paste0("reverselog-", format(base)), trans, inv, 
              log_breaks(base = base), 
              domain = c(1e-100, Inf))
}

modulename <- c("brown", "magenta", "red", "turquoise", "yellow","black", "green", "blue", "pink","grey" )
pltList_host <- list()

#modulename2 <- "blue"
for (ii in modulename) {
  setwd("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/wgcna/GOenrichment")
  #ii <- "red" #modulename[]
  temp  <- ii
  GOplot_wgcna <- read_csv( paste0("Wgcna_GOALL_result_fdr01_",temp ,"_",ver,".csv"))
  #rownames
  GOplot_wgcna <- as.data.frame(GOplot_wgcna)
  rownames(GOplot_wgcna) <- paste0( GOplot_wgcna$Description, " (", GOplot_wgcna$ID,")")
  GOplot_wgcna$Goterm <- try(rownames(GOplot_wgcna),silent=TRUE) 
  # reorder_ Plot GOplot
  GOplot_wgcna2 <- GOplot_wgcna %>%  
                    #top_n(15, wt= -FDR) %>%
                    arrange(#ONTOLOGY, 
                            FDR,
                            -pergene 
                            )  

# filter P-value
gsea_result_Tidy4 <- GOplot_wgcna2 %>% 
  dplyr::select(Goterm, ID, Description, ONTOLOGY, pvalue, FDR, pergene)  %>% 
  filter(FDR< 0.05)

# Plot GOplot
gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(15, wt = -FDR)

#capital letter to small letter
gsea_result_Tidy5$Goterm <- firstup(gsea_result_Tidy5$Goterm)


#facet 0 or 1
#gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001



fig_goplot2 <-       gsea_result_Tidy5      %>%
                    ggplot(aes(x = FDR2,
                               y=reorder(Goterm,-FDR),
                               colour = ONTOLOGY,
                               size = pergene*2)) +
                    geom_point() +
                    #facet_grid( ~ status) +
                    expand_limits(  size= c (0, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_continuous(trans=reverselog_trans(10),
                                       limits = c(0.05, 0.001),
                                       breaks = c(0.05, 0.03, 0.01, 0.001),
                                       labels = c("0.05", "0.03", "0.01",  "<0.001")) +
                    #scale_x_log10()  +
                    #scale_x_reverse(limits = c(0.05, 0.001), 
                    #                breaks = c(0.05, 0.03, 0.01, 0.001),
                    #                labels = c("0.05", "0.03", "0.01",  "<0.001")) +
                    scale_color_brewer(palette="Dark2") +
                    #scale_colour_gradientn(colours = color1 , 
                    #                       space = "Lab",
                    #                       limits=c(0,2.2), 
                    #                       oob = scales::squish,
                    #                       na.value = "#C6DBEF" , 
                    #                       guide=guide_colorbar(reverse=F)) +
                    labs(x=expression(paste("False Discovery Rate")), y="GO term", 
                         colour="Ontology", size="Gene ratio") + 
                    scale_size_continuous(range = c(0, 20), #change the size scale 
                               breaks = c (0.1*2,0.2*2,0.3*2, 0.5*2, 0.6*2, 0.7*2, 0.8*2, 1*2),
                               labels = c("0.1", "0.2","0.3", "0.4", "0.5", "0.6", "0.7", "1")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          #strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=20
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=20 
                                               ), 
                          axis.title.x = element_text(size=20
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=20
                                                 #face="bold",
                                                )
                          )
fig_goplot2
#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(0, "lines"),
                          legend.position="right") 
#paste0 ("Goplot_host_",temp) <- fig_goplot3

# Create plot name.
  pltName <- paste( "Host_",temp, sep = '' )
  pltList_host[[ pltName ]] <- fig_goplot3 


ppi=300
setwd("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/wgcna/GOenrichment/wgcna_GO_plot")
ggsave(paste0 ("Goplot_host_New_",temp,"_",ppi ,"dpi.tiff"), dpi = ppi , height = 10, width = 13) 


#全部をまとめる
library(egg)
ppi=300
tiff(paste0 ("Goplot_host_set_",ppi ,"dpi.tiff"), height = 30*ppi, width = 40*ppi, res=ppi)
par(mar = c(1, 1, 1, 1)); # c(lower ,left, upper, right)
ggpubr::ggarrange(pltList_host$Host_red, pltList_host$Host_brown,pltList_host$Host_magenta,
                  pltList_host$Host_turquoise,pltList_host$Host_yellow, pltList_host$Host_black,
                  pltList_host$Host_black, pltList_host$Host_green, pltList_host$Host_blue,
                  ncol=3, nrow=3,  common.legend = T, legend="bottom", align="hv")
dev.off()
```

##########################
#Microbiome Function
#GO analysis in each module
##########################
#Figures
```{r}
modulename2 <- "Lightcyan3"
modulename <- c("khaki","lightcyan3","royalblue","plum2","salmon", "mediumpurple","violetred","goldenrod","lightgreen"  )
ver <- "1"
pltList_mf <- list()

colors1=c( "#C7E9B4" ,"#C7E9B4" ,"#7FCDBB" ,"#41B6C4" ,"#1D91C0", "#225EA8", "#0C2C84")
color1 <- colorRampPalette(colors = colors1)(300)
colors2=c( " white", "blue")
color2 <- colorRampPalette(colors = colors2)(10)

library("scales")
reverselog_trans <- function(base = exp(1)) {
    trans <- function(x) -log(x, base)
    inv <- function(x) base^(-x)
    trans_new(paste0("reverselog-", format(base)), trans, inv, 
              log_breaks(base = base), 
              domain = c(1e-100, Inf))
}


for (i in modulename) {
  setwd("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/wgcna/GOenrichment")　
  #ii <- "lightcyan3" #modulename[]
  temp  <- i
  GOplot_wgcna <- read_csv( paste0("Wgcna_mbfunc_GOALL_result_fdr01_",temp ,"_",ver,".csv"))
  #rownames
  GOplot_wgcna <- as.data.frame(GOplot_wgcna)
  rownames(GOplot_wgcna) <- paste0( GOplot_wgcna$Description, " (", GOplot_wgcna$ID,")")
  GOplot_wgcna$Goterm <- try(rownames(GOplot_wgcna),silent=TRUE) 
  # reorder_ Plot GOplot
  GOplot_wgcna2 <- GOplot_wgcna %>%  
                    #top_n(15, wt= -FDR) %>%
                    arrange(#ONTOLOGY, 
                             FDR,
                            -pergene)  
# filter P-value
gsea_result_Tidy4 <- GOplot_wgcna2 %>% 
  dplyr::select(Goterm, ID, Description, ONTOLOGY, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 0.05) #%>% filter(pergene > 15)
# Plot GOplot
gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(15, wt = pergene)

#capital letter to small letter
gsea_result_Tidy5$Goterm <- firstup(gsea_result_Tidy5$Goterm)



#facet 0 or 1
#gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")
#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001


fig_goplot2 <-       gsea_result_Tidy5      %>%
                    ggplot(aes(x = FDR2,
                               y=reorder(Goterm,-FDR2),
                               colour = ONTOLOGY,
                               size = pergene*2)) +
                    geom_point() +
                    #facet_grid( ~ status) +
                    expand_limits(  size= c (0, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_continuous(trans=reverselog_trans(10),
                                       limits = c(0.05, 0.001),
                                       breaks = c(0.05, 0.03, 0.01, 0.001),
                                       labels = c("0.05", "0.03", "0.01",  "<0.001")) +
                     scale_color_brewer(palette="Dark2") +
                    labs(title = paste0(temp, " module"), x=expression(paste("False Discovery Rate")), y="GO term", colour="Ontology", size="Gene ratio") + 
                    scale_size_continuous(range = c(0, 20), #change the size scale 
                               breaks = c (0.1*2,0.2*2,0.3*2, 0.5*2, 0.6*2, 0.7*2, 0.8*2, 1*2),
                               labels = c("0.1", "0.2","0.3", "0.4", "0.5", "0.6", "0.7", "1")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          #strip.text.x = element_text(size = 20, color = "black", face = "bold"),
                          axis.text.x = element_text(size=20
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=20 
                                               ), 
                          axis.title.x = element_text(size=20
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=20
                                                 #face="bold",
                                                )
                          )
#fig_goplot2
#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                          legend.position="right") 

fig_goplot3

# Create plot name.
  pltName <- paste( "MF_",temp, sep = '' )
  pltList_mf[[ pltName ]] <- fig_goplot3 

setwd("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/wgcna/GOenrichment/wgcna_GO_plot")
ppi=300
par(mar = c(1, 1, 1, 1)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_mbfun_New_",temp ,"_", ppi ,"dpi.tiff"), height = 10*ppi, width = 15*ppi, res=ppi)
fig_goplot3
dev.off()

ggsave(paste0 ("Goplot_mbfun_New2_",temp,"_",ppi ,"dpi.tiff"), dpi = ppi , height = 15, width = 20) 
}

#全部をまとめる
library(egg)
ppi=300
tiff(paste0 ("Goplot_MF_set_",ppi ,"dpi.tiff"), height = 30*ppi, width = 40*ppi, res=ppi)
par(mar = c(1, 1, 1, 1)); # c(lower ,left, upper, right)
ggpubr::ggarrange(pltList_mf$MF_lightcyan3,pltList_mf$MF_plum2, pltList_mf$MF_salmon,
                  pltList_mf$MF_khaki, pltList_mf$MF_lightgreen,
                  ncol=3, nrow=2,  common.legend = T, legend="bottom", align="hv")
dev.off()
```


##########################
#pcalg
##########################
############################ pcalg:: PC vs. fci, rfci  rsv vs rva#################################

```{r pcalg:: PC }
#https://www.stat.cmu.edu/~cshalizi/uADA/16/lectures/25.pdf
library("pcalg")
library("SMPracticals")
library(dplyr)

# Page 10-
# Get data
data(mathmarks)
head(mathmarks)

# Run PC
suffStat <- list(C=cor(mathmarks),n=nrow(mathmarks))
pc.fit <- pcalg::pc(suffStat, indepTest=gaussCItest, p=ncol(mathmarks), alpha=0.05)
#summary(pc.fit)
# Visualize
library("Rgraphviz")  
Rgraphviz::plot(pc.fit,labels=colnames(mathmarks),main="Inferred DAG for mathmarks")

```

```{r}
#################
MBmodule <- grep("^MB", names(df.all), value = TRUE)
Hostmodule <- grep("^Host", names(df.all), value = TRUE)
TXmodule <- grep("^TX", names(df.all), value = TRUE)
#################
#################Preparation#################
########## DEFINE EXPOSURE and REF LEVEL ##########################
df.all$CPAPintubate<- relevel(as.factor(df.all$CPAPintubate), "0") 
#df_pre$IntensiveTreatment<- relevel(as.factor(df_pre$IntensiveTreatment), "0") 
df_pre <- df.all %>% dplyr::select(c("CPAPintubate", 
                                     "Age_mo",
                                     "intake_sex",
                                     "RSV_NPA",#"RSVanyRV",
                                     #"RSVRVstatus_NPA",
                                     #"RV_CT_NPA",
                                     #"RSV_CT_NPA",  #  "RSVRVstatus_NPA"
                                     all_of(c(Hostmodule, TXmodule, MBmodule)))) #"CPAPintubate"

#df_pre$RSVRVstatus_NPA <- as.factor(df_pre$RSVRVstatus_NPA )
df_pre$RSVanyRV <- as.factor(df_pre$RSVanyRV)
#df_pre$RSVRVstatus_NPA <- as.factor(df_pre$RSVRVstatus_NPA)
df_pre$Y <- ifelse(df.all$CPAPintubate == 2, "Severe", "Nonsevere") %>% as.factor()
#df_models <- df_pre %>% 
#             dplyr::select(-c(CPAPintubate)) 
df_model <- df_pre
dim(df_model)
names(df_model)
```

#Change module name
```{r}
#HF
names(df_model)[which( names(df_model) =="Hostred" ) ] <- "H_Tcellregulation"
names(df_model)[which( names(df_model)=="Hostblue" ) ] <- "H_Neutro_IL1"
names(df_model)[which( names(df_model)=="Hostgreen" ) ] <- "H_GPCR"
names(df_model)[which( names(df_model)=="Hostbrown" ) ] <- "H_TypeI_IFN"
names(df_model)[which( names(df_model)=="Hostturquoise" ) ] <- "H_Ribosome"
#MT
names(df_model)[which( names(df_model)=="TXdarkred" ) ] <- "pneumoniae_aureus"
names(df_model)[which( names(df_model)=="TXdarkorchid" ) ] <- "Mixed_1"
names(df_model)[which( names(df_model)=="TXdarkturquoise" ) ] <- "Moraxella"
names(df_model)[which( names(df_model)=="TXdarkblue" ) ] <- "Streptcoccus"
names(df_model)[which( names(df_model)=="TXdarkorange" ) ] <- "Haemophilus"
#MF
names(df_model)[which( names(df_model)=="MBlightcyan3" ) ] <- "M_Plasma_membrane"
names(df_model)[which( names(df_model)=="MBplum2" ) ] <- "M_mRNA"
names(df_model)[which( names(df_model)=="MBroyalblue" ) ] <- "M_BCAA"
names(df_model)[which( names(df_model)=="MBsalmon" ) ] <- "M_Oxidative_stress"
names(df_model)[which( names(df_model)=="MBmediumpurple" ) ] <- "M_NADH"

names(df_model)
```

#Use Real data – upload the data
```{r DAG real data}
set.seed(111)
library("pcalg")
library("tidyverse")


for (i in 1:20) {
  df_model[,i] <- as.numeric(df_model[,i])
}

#df_model$RSVanyRV <- as.factor(df_model$RSVanyRV)
df_model$intake_sex <- as.factor(df_model$intake_sex)

names(df_model)



#df_pcalg <- df_model[,c( 2,12,4,14,10,13,7,18,17,16,6,8,15,5,1)]# 9,11.,19 c( 2,12,4,14,10,13,7,18,17,16,6,8,15,5,1)
#df_pcalg <- df_model[,c(14,10,12,13,18,11,16,19,15,17,9,8,7,6,5,1)]
df_pcalg <- df_model[,c(14,10,12,13,18,16,19,15,17,8,7,6,5,1)]

#df_pcalg <- r.autoscale(df_pcalg)

# Run PC
suffStat <- list(C=cor(df_pcalg ),n=nrow(df_pcalg ))

pc.fit <- pcalg::pc(suffStat, 
                    indepTest=gaussCItest,  # Gaussian data (gaussCItest), discrete data (disCItest), and binary data (see binCItest).
                    p=ncol(df_pcalg), 
                    skel.method = "stable.fast", #, "stable.fast", "stable","original"
                    #u2pd = "relaxed",#c("relaxed", "rand", "retry"),
                    alpha=0.37,
                    NAdelete = T,
                    verbose=F,  # detailed output is provided.
                    conservative = F,
                    #maj.rule = T,
                    solve.confl = F)

#summary(pc.fit)

#Visualization
dev.off()
plot_dag <- Rgraphviz::plot(pc.fit , 
                labels=colnames(df_pcalg), 
                main="Inferred DAG in PPV use")


tiff("Inferred_DAG_PPVuse_1006_.tiff", height = 1000, width = 1200)
plot_dag
dev.off()
#plot_dag <- iplotPC(pc.fit,labels=colnames(df_pcalg))
```

#Preservation
```{r }
library("Rgraphviz")

tiff("Inferred_DAG_PPVuse_1005_.tiff", height = 1000, width = 1200)
plot_dag
dev.off()

```



#############################
Figure 2. Differential gene expression analysis of host transcriptome data with regard to the use of positive pressure ventilation in infants hospitalized for bronchiolitis 
###########
A)	Volcano plot of differentially expressed microbial genes (metatranscriptome).
###########
#データをDeSeq2でもってこないとできないので、もとデータにもどること


#############################
B)	Gene set enrichment analysis (transcriptome). We showed 30 host pathways (GO biological process) with the highest absolute normalized enrichment score in the gene set enrichment analysis (GSEA) with down-regulated pathways on the left side and up-regulated pathways on the right side. We also showed the normalized enrichment score and the gene ratio for the corresponding pathways. 
#############################


##############################
B) pathway analysis in Figure 2
Biological process
#############################

# Gene ratio にてselection
```{r}
#load data
gsea_result_df <- read.csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/finaldata/GSEA_GOBP_result_full.csv")
gsea_result_df$Description[which(gsea_result_df$Description =="Rna splicing, via transesterification reactions with bulged adenosine as nucleophile")] <- "RNA splicing, via transesterification reactions with bulged adenosine as nucleophile"

N <- dim(gsea_result_df)[1]
#Core enrichmentの数を数える
for (i in 1:N) {
 gsea_result_df$Count[i] <- length(str_extract_all(gsea_result_df$core_enrichment[i], pattern="/", simplify=TRUE)) +1 
}
gsea_result_df$pergene <-  (gsea_result_df$Count/gsea_result_df$setSize)

#Add ID
#gsea_result_df$ID <- rownames(gsea_result_df)
rownames(gsea_result_df) <- paste0(gsea_result_df$Description, " (",gsea_result_df$ID, ")" )
gsea_result_df$Goterm <-    rownames(gsea_result_df)  

  
# filter gene percent
gsea_result_df2 <- gsea_result_df %>% 
  dplyr::select(Goterm, ID, Description, NES, pvalue, p.adjust, pergene)  %>% 
  filter(pergene >= 0.05)

#calculate FDR
gsea_result_df2$FDR <- p.adjust(gsea_result_df2$pvalue, method = "fdr")


gsea_result_df3 <- gsea_result_df2 %>% filter(gsea_result_df2$FDR < 0.15)
write.csv(gsea_result_df, file = "GSEA_GO_result_ALL.csv", row.names = FALSE)
write.csv(gsea_result_df3, file = "GSEA_GO_result_p015.csv", row.names = FALSE)

gsea_result_df4 <- gsea_result_df2 %>% filter(gsea_result_df2$FDR < 0.05)
```

# Show in a tabl
```{r}
library(DT)
# Tidy the results:
gsea_result_Tidy <- gsea_result_df2 %>%
  as_tibble() %>%
  arrange(desc(NES))

gsea_result_Tidy$absNES <- abs(gsea_result_Tidy$NES)

gsea_result_Tidy2 <- gsea_result_Tidy %>% top_n(30, wt=absNES)

gsea_result_Tidy3 <- gsea_result_Tidy %>% top_n(30, wt=-p.adjust)

# Show in a table - adjust for fgsea by adding dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
gsea_result_Tidy3  %>% 
  #top_n(15, wt= abs(NES)) %>%
  arrange(FDR) %>% 
  DT::datatable()
```

# axis label name color https://stackoverflow.com/questions/38862303/customize-ggplot2-axis-labels-with-different-colors
#Plot GO pathway
#NES sort　figurefigure
```{r}
colors1=c( "#C7E9B4" ,"#C7E9B4" ,"#7FCDBB" ,"#41B6C4" ,"#1D91C0", "#225EA8", "#0C2C84")
colors11=c( "#0C2C84", "#225EA8","#1D91C0" ,"#41B6C4","#7FCDBB","#C7E9B4","#C7E9B4")

color1 <- colorRampPalette(colors = colors1)(100)

library(circlize)
#colors3=colorRamp2(c(0.5,0.75, 1.0,1.2, 1.4, 1.45,1.5,1.55,1.6,1.65,1.7,1.75,1.8,1.85,1.9,2.0,2.05,2.1,2.15,2.2, 2.25,2.3,2.35,2.4,2.5),color1 )


# filter P-value
gsea_result_Tidy4 <- gsea_result_Tidy %>% 
  dplyr::select(Goterm, ID, Description, NES, absNES, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 0.1)

# Plot GOplot
gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(30, wt = absNES)
#gsea_result_Tidy5 <- gsea_result_Tidy %>% top_n(30, wt=-p.adjust)

#facet 0 or 1
gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001


fig_goplot2 <-       gsea_result_Tidy5      %>%
                    ggplot(aes(x = FDR2,
                               y=reorder(Goterm,-FDR),
                               colour = absNES,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_reverse(limits = c(0.1, 0.001), 
                                    breaks = c(0.1, 0.05, 0.03, 0.01, 0.001),
                                    labels = c("0.1","0.05", "0.03", "0.01",  "≤0.001")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.5,2.5), 
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=F)) +
                    labs(x=expression(paste("False Discovery Rate")), y="GO term", colour="Normalized Enrichment Scores", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), #change the size scale 
                               breaks = c (0.3, 0.4, 0.5, 0.6, 0.7),
                               labels = c( "0.3", "0.4", "0.5", "0.6", "0.7")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=13
                                                 #face="bold",
                                                )
                          )
fig_goplot2


#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                          legend.position="bottom", 
                      #legend.box="vertical"
                      ) 


fig_goplot3

ppi=300
par(mar = c(2, 2, 2, 2)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_hostBP1",ppi ,"dpi.tiff"), height = 11*ppi, width = 20*ppi, res=ppi)
fig_goplot3
dev.off()
```

#Plot GO pathway
#これつかう　figurefigure
# p-value sort
```{r}
colors1=c( "#C7E9B4" ,"#C7E9B4" ,"#7FCDBB" ,"#41B6C4" ,"#1D91C0", "#225EA8", "#0C2C84")
colors11=c( "#0C2C84","#7FCDBB","#C7E9B4","#C7E9B4")

color1 <- colorRampPalette(colors = colors11)(10)

library(circlize)
colors3=colorRamp2(c(0.5,0.75, 1.0,1.2, 1.4, 1.45,1.5,1.55,1.6,1.65,1.7,1.75,1.8,1.85,1.9,2.0,2.05,2.1,2.15,2.2, 2.25,2.3,2.35,2.4,2.5),color1 )


# filter P-value
gsea_result_Tidy4 <- gsea_result_Tidy %>% 
  dplyr::select(Goterm, ID, Description, NES, absNES, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 0.1)

# Plot GOplot
#gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(30, wt = absNES)
gsea_result_Tidy5 <- gsea_result_Tidy %>% top_n(30, wt=-FDR)

#facet 0 or 1
gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001


fig_goplot2 <-       gsea_result_Tidy5      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05","≥0.10"), #≤≥
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", colour="FDR", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), #change the size scale 
                               breaks = c (0.2, 0.3, 0.4, 0.5, 0.6),
                               labels = c( "0.2", "0.3", "0.4", "0.5", "0.6")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=13
                                                 #face="bold",
                                                )
                          )
fig_goplot2


#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                          legend.position="right"#, 
                      #legend.box="vertical"
                      ) 


fig_goplot3

ppi=300
par(mar = c(0.1, 0.1, 0.1, 0.1)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_hostBP_1007_",ppi ,"dpi.tiff"), height = 11*ppi, width = 20*ppi, res=ppi)
fig_goplot3
dev.off()
```

##############################
Celluar component
Supplemental file Ex
##############################

# Gene ratio にてselection
```{r}
#load data
gsea_result_df <- read.csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/finaldata/GSEA_GOCC_result.csv")

N <- dim(gsea_result_df)[1]
#Core enrichmentの数を数える
for (i in 1:N) {
 gsea_result_df$Count[i] <- length(str_extract_all(gsea_result_df$core_enrichment[i], pattern="/", simplify=TRUE)) +1 
}
gsea_result_df$pergene <-  (gsea_result_df$Count/gsea_result_df$setSize)


#Add ID
#gsea_result_df$ID <- rownames(gsea_result_df)
rownames(gsea_result_df) <- paste0(gsea_result_df$Description, " (",gsea_result_df$ID, ")" )
gsea_result_df$Goterm <-    rownames(gsea_result_df)  

  
# filter gene percent
gsea_result_df2 <- gsea_result_df %>% 
  dplyr::select(Goterm, ID, Description, NES, pvalue, p.adjust, pergene)  %>% 
  filter(pergene >= 0.05)

#calculate FDR
gsea_result_df2$FDR <- p.adjust(gsea_result_df2$pvalue, method = "fdr")


gsea_result_df3 <- gsea_result_df2 %>% filter(gsea_result_df2$FDR < 0.15)
write.csv(gsea_result_df, file = "GSEA_GO_result_MF_ALL.csv", row.names = FALSE)
write.csv(gsea_result_df3, file = "GSEA_GO_result_MF_p015.csv", row.names = FALSE)
```


# Show in a tabl
```{r}
# Tidy the results:
gsea_result_Tidy <- gsea_result_df2 %>%
  as_tibble() %>%
  arrange(desc(NES))

gsea_result_Tidy$absNES <- abs(gsea_result_Tidy$NES)

gsea_result_Tidy2 <- gsea_result_Tidy %>% top_n(30, wt=absNES)

# Show in a table - adjust for fgsea by adding dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
gsea_result_Tidy2  %>% 
  #top_n(15, wt= abs(NES)) %>%
  arrange(FDR) %>% 
  DT::datatable()
```

# axis label name color 
https://stackoverflow.com/questions/38862303/customize-ggplot2-axis-labels-with-different-colors
#Plot GO pathway
# TOP NES
```{r}
#colors1=c( "#C7E9B4" ,"#C7E9B4" ,"#7FCDBB" ,"#41B6C4" ,"#1D91C0", "#225EA8", "#0C2C84")
#color1 <- colorRampPalette(colors = colors1)(25)

library(circlize)
colors3=colorRamp2(c(0.5,0.75, 1.0,1.2, 1.4, 1.45,1.5,1.55,1.6,1.65,1.7,1.75,1.8,1.85,1.9,2.0,2.05,2.1,2.15,2.2, 2.25,2.3,2.35,2.4,2.5),color1 )


# filter P-value
gsea_result_Tidy4 <- gsea_result_Tidy %>% 
  dplyr::select(Goterm, ID, Description, NES, absNES, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 0.1)

# Plot GOplot
gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(30, wt = absNES)

#facet 0 or 1
gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001


fig_goplot2 <-       gsea_result_Tidy5      %>%
                    ggplot(aes(x = FDR2,
                               y=reorder(Goterm,-FDR),
                               colour = absNES,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_reverse(limits = c(0.1, 0.001), 
                                    breaks = c(0.1, 0.05, 0.03, 0.01, 0.001),
                                    labels = c("0.1","0.05", "0.03", "0.01",  "<0.001")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.5,2.5), 
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=F)) +
                    labs(x=expression(paste("False Discovery Rate")), y="GO term", colour="Normalized Enrichment Scores", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), #change the size scale 
                               breaks = c (0.3, 0.4, 0.5, 0.6, 0.7),
                               labels = c( "0.3", "0.4", "0.5", "0.6", "0.7")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=13
                                                 #face="bold",
                                                )
                          )
fig_goplot2


#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                          legend.position="bottom", 
                      legend.box="vertical") 


fig_goplot3

ppi=300
par(mar = c(2, 2, 2, 2)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_host_CC_1007_",ppi ,"dpi.tiff"), height = 10*ppi, width = 18*ppi, res=ppi)
fig_goplot3
dev.off()
```


#Plot GO pathway
#これつかう　figurefigure
# p-value sort
```{r}
colors1=c( "#C7E9B4" ,"#C7E9B4" ,"#7FCDBB" ,"#41B6C4" ,"#1D91C0", "#225EA8", "#0C2C84")
colors11=c( "#0C2C84","#7FCDBB","#C7E9B4","#C7E9B4")

color1 <- colorRampPalette(colors = colors11)(1000)

library(circlize)
colors3=colorRamp2(c(0.5,0.75, 1.0,1.2, 1.4, 1.45,1.5,1.55,1.6,1.65,1.7,1.75,1.8,1.85,1.9,2.0,2.05,2.1,2.15,2.2, 2.25,2.3,2.35,2.4,2.5),color1 )


# filter P-value
gsea_result_Tidy4 <- gsea_result_Tidy %>% 
  dplyr::select(Goterm, ID, Description, NES, absNES, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 0.1)

# Plot GOplot
#gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(30, wt = absNES)
gsea_result_Tidy5 <- gsea_result_Tidy %>% top_n(30, wt=-FDR)

#facet 0 or 1
gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001
```


```{r}
#Upper  lower alphabet
#gsea_result_Tidy5$Goterm <- str_to_sentence(gsea_result_Tidy5$Goterm, locale = "en")
#個別に修正
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Mhc protein complex (GO:0042611)")] <- "MHC protein complex (GO:0042611)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Mhc class ii protein complex (GO:0042613)")] <- "MHC class II protein complex (GO:0042613)"
```


```{r}
colors=c( "#0C2C84","#7FCDBB","#C7E9B4","#C7E9B4")
color1 <- colorRampPalette(colors = colors)(1000)

write.csv(gsea_result_Tidy5, "data_object_figS3B.csv", row.names = F)
df_figS3B <- read.csv("data_object_figS3B.csv")



fig_S3B <-       df_figS3B       %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                     scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05","≥0.10"), # 
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", size="Gene ratio", colour="FDR") + 
                    scale_size_continuous(range = c(3, 10), #change the size scale 
                               breaks = c (0.3, 0.4, 0.5, 0.6, 0.7),
                               labels = c( "0.3", "0.4", "0.5", "0.6", "0.7")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=13
                                                 #face="bold",
                                                )
                          )
fig_goplot2


#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                          legend.position="right"#, 
                      #legend.box="vertical"
                      ) 


ppi=300
par(mar = c(2, 2, 2, 2)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_hostCC_psort_r1.tiff"), height = 11*ppi, width = 20*ppi, res=ppi)
fig_goplot3
dev.off()
```

############################################
#Molecular Function
Supplemental file Ex
###########################################

# Gene ratio にてselection
```{r}
#load data
gsea_result_df <- read.csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/finaldata/GSEA_GOMF_result.csv")

N <- dim(gsea_result_df)[1]
#Core enrichmentの数を数える
for (i in 1:N) {
 gsea_result_df$Count[i] <- length(str_extract_all(gsea_result_df$core_enrichment[i], pattern="/", simplify=TRUE)) +1 
}
gsea_result_df$pergene <-  (gsea_result_df$Count/gsea_result_df$setSize)


#Add ID
#gsea_result_df$ID <- rownames(gsea_result_df)
rownames(gsea_result_df) <- paste0(gsea_result_df$Description, " (",gsea_result_df$ID, ")" )
gsea_result_df$Goterm <-    rownames(gsea_result_df)  

  
# filter gene percent
gsea_result_df2 <- gsea_result_df %>% 
  dplyr::select(Goterm, ID, Description, NES, pvalue, p.adjust, pergene)  %>% 
  filter(pergene >= 0.05)

#calculate FDR
gsea_result_df2$FDR <- p.adjust(gsea_result_df2$pvalue, method = "fdr")


gsea_result_df3 <- gsea_result_df2 %>% filter(gsea_result_df2$FDR < 0.15)
write.csv(gsea_result_df, file = "GSEA_GO_result_MF_ALL.csv", row.names = FALSE)
write.csv(gsea_result_df3, file = "GSEA_GO_result_MF_p015.csv", row.names = FALSE)
```

# Show in a tabl
```{r}
# Tidy the results:
gsea_result_Tidy <- gsea_result_df2 %>%
  as_tibble() %>%
  arrange(desc(NES))

gsea_result_Tidy$absNES <- abs(gsea_result_Tidy$NES)

gsea_result_Tidy2 <- gsea_result_Tidy %>% top_n(30, wt=absNES)

# Show in a table - adjust for fgsea by adding dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
gsea_result_Tidy2  %>% 
  #top_n(15, wt= abs(NES)) %>%
  arrange(FDR) %>% 
  DT::datatable()
```

# axis label name color 
https://stackoverflow.com/questions/38862303/customize-ggplot2-axis-labels-with-different-colors
#Plot GO pathway
#これつかう　figurefigure
# TOP NES
```{r}
colors1=c( "#C7E9B4" ,"#C7E9B4" ,"#7FCDBB" ,"#41B6C4" ,"#1D91C0", "#225EA8", "#0C2C84")
color1 <- colorRampPalette(colors = colors1)(25)

library(circlize)
colors3=colorRamp2(c(0.5,0.75, 1.0,1.2, 1.4, 1.45,1.5,1.55,1.6,1.65,1.7,1.75,1.8,1.85,1.9,2.0,2.05,2.1,2.15,2.2, 2.25,2.3,2.35,2.4,2.5),color1 )


# filter P-value
gsea_result_Tidy4 <- gsea_result_Tidy %>% 
  dplyr::select(Goterm, ID, Description, NES, absNES, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 0.1)

# Plot GOplot
gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(30, wt = absNES)

#facet 0 or 1
gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001


fig_goplot2 <-       gsea_result_Tidy5      %>%
                    ggplot(aes(x = FDR2,
                               y=reorder(Goterm,-FDR),
                               colour = absNES,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_reverse(limits = c(0.1, 0.001), 
                                    breaks = c(0.1, 0.05, 0.03, 0.01, 0.001),
                                    labels = c("0.1","0.05", "0.03", "0.01",  "<0.001")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.5,2.5), 
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=F)) +
                    labs(x=expression(paste("False Discovery Rate")), y="GO term", colour="Normalized Enrichment Scores", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), #change the size scale 
                               breaks = c (0.3, 0.4, 0.5, 0.6, 0.7),
                               labels = c( "0.3", "0.4", "0.5", "0.6", "0.7")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=13
                                                 #face="bold",
                                                )
                          )
fig_goplot2


#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                          legend.position="bottom", 
                      legend.box="vertical") 


fig_goplot3

ppi=300
par(mar = c(2, 2, 2, 2)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_host_CC1_",ppi ,"dpi.tiff"), height = 10*ppi, width = 18*ppi, res=ppi)
fig_goplot3
dev.off()
```

#Plot GO pathway
#これつかう　figurefigure
# p-value sort
```{r}
colors1=c( "#C7E9B4" ,"#C7E9B4" ,"#7FCDBB" ,"#41B6C4" ,"#1D91C0", "#225EA8", "#0C2C84")
colors11=c( "#0C2C84","#7FCDBB","#C7E9B4","#C7E9B4")

color1 <- colorRampPalette(colors = colors11)(1000)

library(circlize)
colors3=colorRamp2(c(0.5,0.75, 1.0,1.2, 1.4, 1.45,1.5,1.55,1.6,1.65,1.7,1.75,1.8,1.85,1.9,2.0,2.05,2.1,2.15,2.2, 2.25,2.3,2.35,2.4,2.5),color1 )


# filter P-value
gsea_result_Tidy4 <- gsea_result_Tidy %>% 
  dplyr::select(Goterm, ID, Description, NES, absNES, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 0.1)

# Plot GOplot
#gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(30, wt = absNES)
gsea_result_Tidy5 <- gsea_result_Tidy %>% top_n(30, wt=-FDR)

#facet 0 or 1
gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001
```

```{r}
#Upper  lower alphabet
#gsea_result_Tidy5$Goterm <- str_to_sentence(gsea_result_Tidy5$Goterm, locale = "en")
#個別に修正
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Rrna binding (GO:0019843)")] <- "rRNA binding (GO:0019843)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Nadh dehydrogenase (quinone) activity (GO:0050136)")] <- "NADH dehydrogenase (quinone) activity (GO:0050136)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Nadh dehydrogenase (ubiquinone) activity (GO:0008137)")] <- "NADH dehydrogenase (ubiquinone) activity (GO:0008137)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Nadh dehydrogenase activity (GO:0003954)")] <- "NADH dehydrogenase activity (GO:0003954)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Nadh dehydrogenase activity (GO:0003954)")] <- "NADH dehydrogenase activity (GO:0003954)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="")] <- ""
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="")] <- ""
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="")] <- ""
```

```{r}
colors=c( "#0C2C84","#7FCDBB","#C7E9B4","#C7E9B4")
color1 <- colorRampPalette(colors = colors)(1000)

write.csv(gsea_result_Tidy5, "data_object_figS3A.csv", row.names = F)
df_figS3A <- read.csv("data_object_figS3A.csv")

fig_goplot2 <-      df_figS3A      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05","≥0.10"), #≤≥
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", colour="FDR", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), #change the size scale 
                               breaks = c (0.2, 0.3, 0.4, 0.5, 0.6),
                               labels = c( "0.2", "0.3", "0.4", "0.5", "0.6")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=13
                                                 #face="bold",
                                                )
                          )
fig_goplot2


#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                          legend.position="right"#, 
                      #legend.box="vertical"
                      ) 


fig_goplot3


ppi=300
par(mar = c(2, 2, 2, 2)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_hostMF_psort_r1.tiff"), height = 11*ppi, width = 20*ppi, res=ppi)
fig_goplot3
dev.off()
```


###########
Figure 4. Differential gene expression analysis of microbial function data with regard to the use of positive pressure ventilation in infants hospitalized for bronchiolitis 
###########
A)	Volcano plot of differentially expressed microbial genes (metatranscriptome).
###########
```{r}
#load data
ttest_func <- read.csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/finaldata/ttest_func_lastdata2_fi.csv")

sum((ttest_func$p_adjust < 0.05 & abs(ttest_func$log2FC) >= 0.58), na.rm=TRUE)


#Volcano
df_volcano <- ttest_func[, c(1,5,7)]

rownames(df_volcano) <- df_volcano$name
#names(df_volcano)[which( names(df_volcano) =="p_adjust" ) ] <- "adjusted P value"

df_volcano$p_adjust[which( df_volcano$p_adjust == 0.0000 ) ] <- 0.0001
head(df_volcano ,10)


#Add label
keyvals <- ifelse( df_volcano$log2FC < -0.58 & df_volcano$p_adjust <0.05 , "royalblue",
                            ifelse(df_volcano$log2FC > 0.58 & df_volcano$p_adjust <0.05, "red2",
                                  ifelse(abs(df_volcano$log2FC) >= 0.58 & df_volcano$p_adjust >= 0.05, "forestgreen",
                                      ifelse(abs(df_volcano$log2FC) < 0.58 & df_volcano$p_adjust < 0.05, "#FFAB40", "grey30"))))

levels(keyvals) <- c("grey30","red2","forestgreen","#FFAB40","royalblue" )


  keyvals[is.na(keyvals)] <- "grey30"
  names(keyvals)[keyvals == "red2"] <- "Upregulated*"
  names(keyvals)[keyvals == "grey30"] <- "Not significant"
  names(keyvals)[keyvals == "forestgreen"] <- "Log FC ≥|0.58|"
  names(keyvals)[keyvals == "#FFAB40"] <- "FDR<0.05"
  names(keyvals)[keyvals == "royalblue"] <- "Downregulated*"
  
levels(keyvals) <- c("Upregulated","Downregulated","Log FC ≥|0.58|","FDR<0.05","Not significant" )


#volcano
library(EnhancedVolcano)
volcano <- EnhancedVolcano(df_volcano,
                  lab = rownames(df_volcano),
                  x = 'log2FC',
                  y = 'p_adjust',
                  xlim = c(-10, 10),
                  pCutoff = 0.05,
                  FCcutoff = 1  ##log2(0.58) = FC1.5; log2(1)= FC2
                  )
volcano

volcano2 <- EnhancedVolcano::EnhancedVolcano(df_volcano,
    lab =  rownames(df_volcano) ,               #rownames(res_sub),
    x = 'log2FC',
    y = "p_adjust", #'pvalue',
    selectLab = c(""),
    #xlim = c(-10, 10),
    #title = '2 versus 1',
    pCutoff = 0.05,
    FCcutoff = 0.58, #log2(0.58) = FC1.5; log2(1)= FC2
    #legendLabels=c('Not sig.','Log (base 2) FC','p-value','p-value & Log (base 2) FC'),
    pointSize = 1,
    #legendLabels=c(expression(paste('No significant')),
    #               expression(paste("Log"[2]," fold change")),
    #               expression(paste('P-value')), 
    #               expression(paste("Log"[2]," fold change"," &", " P-value "))) ,
    legendPosition = "right",
    colCustom = keyvals,
    #boxedLabels = TRUE
    #drawConnectors = TRUE,
    #widthConnectors = 0.3,
    #colConnectors = 'grey50',
    #arrowheads = FALSE,
    labSize = 3, 
    title = NULL, #"",
    subtitle = NULL, #"",
    caption = NULL, #paste0("total = ", nrow(toptable), " variables"),
    xlab = bquote(~Log[2]~ 'fold change'),
    ylab = bquote(~-Log[10]~ 'FDR'),
    )



volcano3 <- volcano2 + 
  ggplot2::coord_cartesian(xlim=c(-5, 5),ylim=c(0, 8)) + 
  scale_x_continuous(breaks = c(-7,-5,-3,-1,0,1,3,5,7))+ 
  scale_y_continuous(breaks = c(0, 2.5, 5.0, 7.5))

volcano3

ppi=300
par(mar = c(1, 1, 1, 1)); # c(lower ,left, upper, right)
tiff(paste0 ("Volcano_CPAP_func_r1.tiff"), height = 6*ppi, width = 10*ppi, res=ppi)
volcano3
dev.off()
```


##################
B) pathway analysis
##################
Biological Process
#################
#GO-GSEA by clusterProfiler gseGO using ttest data (t-statistics)
```{r}
library(clusterProfiler)
library(DOSE)
library(org.Hs.eg.db)
library(ggplot2)
library(dplyr)
library("AnnotationDbi")
library("org.Hs.eg.db")
library("org.EcK12.eg.db")
library("org.EcSakai.eg.db")


data(geneList)
#Prepare Input
# we want the log2 fold change 
#original_gene_list <- res$log2FoldChange
# we want the p-value

ttest_func$ENTREZID = mapIds(org.EcK12.eg.db,
                    keys=  ttest_func$name ,
                    column="ENTREZID",
                    keytype ="SYMBOL",
                    multiVals="first")

#fullgene_list <- res_ttest  %>% filter(!(is.na(res_ttest$p_adjust)))
fullgene_list <- ttest_func %>% filter(!(is.na(ttest_func$log2FC)))


#original_gene_list <- fullgene_list$p_adjust
original_gene_list <- fullgene_list$log2FC
# we want the p-value
#original_gene_list <- res$stat
# name the vector
#names(original_gene_list) <- fullgene_list$ENTREZID
names(original_gene_list) <- fullgene_list$name

original_gene_list_df <- as.data.frame(original_gene_list)
original_gene_list_df <- data.frame(fullgene_list$ENTREZID, original_gene_list_df )

#gene_list <- left_join(original_gene_list_df, all.genes.entrez, by = c("rownames.original_gene_list_df." = "ENSEMBL"))
original_gene_list_df <- original_gene_list_df %>% filter(!is.na(original_gene_list_df))

#original_gene_list_df <- original_gene_list_df %>% distinct(fullgene_list.ENTREZID)

## step 0
gene_list_ori <- original_gene_list_df
## Step 1: numeric vector
gene_list <- gene_list_ori[,2]
## Step 2: named vector
names(gene_list) = as.character(gene_list_ori[,1])
## Step 3: decreasing order
gene_list = sort(gene_list, decreasing = TRUE)


#開始時と終了時の差を取る 
start.time<-proc.time() #開始
#GSEA
gse_result <- clusterProfiler::gseGO(geneList     = gene_list ,  # 全遺伝子,
                    OrgDb        = org.EcK12.eg.db,
                    ont          = "BP", # "BP","CC","MF","ALL"から選択
                    #nPerm        = 1000,
                    #nPermSimple = 1000,
                    pvalueCutoff = 1,
                    #qvalueCutoff = 0.15,
                    minGSSize = 10,
                    maxGSSize = 500, 
                    eps = 0, 
                    pAdjustMethod = "fdr", 
                    verbose = TRUE, 
                    seed = 622,
                    by = "fgsea"
                    #pool = T
                    )

gsea_result_df <- as.data.frame(gse_result)
#View(gsea_result_df) 
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

```{r}
gsea_result_df <- as.data.frame(gse_result)
#Upper  lower alphabet
gsea_result_df$Description <- str_to_sentence(gsea_result_df$Description, locale = "en")

#個別に修正
#gsea_result_df$Description[which(gsea_result_df$Description =="Positive t cell selection")] <- "Positive T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Cellular response to type i interferon")] <- "Cellular response to type I interferon"
#gsea_result_df$Description[which(gsea_result_df$Description =="Type i interferon signaling pathway")] <- "Type I interferon signaling pathway"
#gsea_result_df$Description[which(gsea_result_df$Description =="Thymic t cell selection")] <- "Thymic T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Thymic t cell selection")] <- "Thymic T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Positive t cell selection")] <- "Positive T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Response to type i interferon")] <- "Response to type I interferon"
#gsea_result_df$Description[which(gsea_result_df$Description =="")] <- ""

#GSEAの可視化
library(ggplot2)
library(ggridges)
#ridgeplot(gse_result,showCategory = 8)

write.csv(gsea_result_df, file = "GSEA_GO_result_microfunc_BP_r1.csv", row.names = FALSE)

```

#GO-GSEA by clusterProfiler gseGO using ttest data (t-statistics)
```{r}
library(clusterProfiler)
library(DOSE)
library(org.Hs.eg.db)
library(ggplot2)
library(dplyr)
library("AnnotationDbi")
library("org.Hs.eg.db")
library("org.EcK12.eg.db")
library("org.EcSakai.eg.db")


data(geneList)
#Prepare Input
# we want the log2 fold change 
#original_gene_list <- res$log2FoldChange
# we want the p-value

ttest_func$ENTREZID = mapIds(org.EcK12.eg.db,
                    keys=  ttest_func$name ,
                    column="ENTREZID",
                    keytype ="SYMBOL",
                    multiVals="first")

#fullgene_list <- res_ttest  %>% filter(!(is.na(res_ttest$p_adjust)))
fullgene_list <- ttest_func %>% filter(!(is.na(ttest_func$stat)))


#original_gene_list <- fullgene_list$p_adjust
original_gene_list <- fullgene_list$stat
# we want the p-value
#original_gene_list <- res$stat
# name the vector
#names(original_gene_list) <- fullgene_list$ENTREZID
names(original_gene_list) <- fullgene_list$name

original_gene_list_df <- as.data.frame(original_gene_list)
original_gene_list_df <- data.frame(fullgene_list$ENTREZID, original_gene_list_df )

#gene_list <- left_join(original_gene_list_df, all.genes.entrez, by = c("rownames.original_gene_list_df." = "ENSEMBL"))
original_gene_list_df <- original_gene_list_df %>% filter(!is.na(original_gene_list_df))

#original_gene_list_df <- original_gene_list_df %>% distinct(fullgene_list.ENTREZID)

## step 0
gene_list_ori <- original_gene_list_df
## Step 1: numeric vector
gene_list <- gene_list_ori[,2]
## Step 2: named vector
names(gene_list) = as.character(gene_list_ori[,1])
## Step 3: decreasing order
gene_list = sort(gene_list, decreasing = TRUE)


#開始時と終了時の差を取る 
start.time<-proc.time() #開始
#GSEA
gse_result <- clusterProfiler::gseGO(geneList     = gene_list ,  # 全遺伝子,
                    OrgDb        = org.EcK12.eg.db,
                    ont          = "BP", # "BP","CC","MF","ALL"から選択
                    #nPerm        = 1000,
                    nPermSimple = 1000,
                    pvalueCutoff = 1,
                    #qvalueCutoff = 0.15,
                    minGSSize = 10,
                    maxGSSize = 500, 
                    eps = 0, 
                    pAdjustMethod = "fdr", 
                    verbose = TRUE, 
                    seed = 622,
                    by = "fgsea"
                    #pool = T
                    )

gsea_result_df <- as.data.frame(gse_result)
#View(gsea_result_df) 
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

```{r}
gsea_result_df <- as.data.frame(gse_result)
#Upper  lower alphabet
gsea_result_df$Description <- str_to_sentence(gsea_result_df$Description, locale = "en")

#個別に修正
#gsea_result_df$Description[which(gsea_result_df$Description =="Positive t cell selection")] <- "Positive T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Cellular response to type i interferon")] <- "Cellular response to type I interferon"
#gsea_result_df$Description[which(gsea_result_df$Description =="Type i interferon signaling pathway")] <- "Type I interferon signaling pathway"
#gsea_result_df$Description[which(gsea_result_df$Description =="Thymic t cell selection")] <- "Thymic T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Thymic t cell selection")] <- "Thymic T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Positive t cell selection")] <- "Positive T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Response to type i interferon")] <- "Response to type I interferon"
#gsea_result_df$Description[which(gsea_result_df$Description =="")] <- ""

#GSEAの可視化
library(ggplot2)
library(ggridges)
#ridgeplot(gse_result,showCategory = 8)

write.csv(gsea_result_df, file = "GSEA_GO_result_microfunc_BP.csv", row.names = FALSE)

```

# Gene ratio にてselection
```{r}
N <- dim(gsea_result_df)[1]
#Core enrichmentの数を数える
for (i in 1:N) {
 gsea_result_df$Count[i] <- length(str_extract_all(gsea_result_df$core_enrichment[i], pattern="/", simplify=TRUE)) +1 
}
gsea_result_df$pergene <-  (gsea_result_df$Count/gsea_result_df$setSize)


#Add ID
gsea_result_df$ID <- rownames(gsea_result_df)
rownames(gsea_result_df) <- paste0(gsea_result_df$Description, " (",gsea_result_df$ID, ")" )
gsea_result_df$Goterm <-    rownames(gsea_result_df)  

  
# filter gene percent
gsea_result_df2 <- gsea_result_df %>% 
  dplyr::select(Goterm, ID, Description, NES, pvalue, p.adjust, pergene)  %>% 
  filter(pergene >= 0.05)

#calculate FDR
gsea_result_df2$FDR <- p.adjust(gsea_result_df2$pvalue, method = "fdr")
gsea_result_df3 <- gsea_result_df2 %>% filter(gsea_result_df2$FDR < 0.50)


write.csv(gsea_result_df, file = "GSEA_GO_result_microfuncBP_FC_ALL.csv", row.names = FALSE)
write.csv(gsea_result_df3, file = "GSEA_GO_result_microfuncBP_FC_p05.csv", row.names = FALSE)
```
###############################
#Visualization
#Figurefigure
###############################
# Show in a tabl
```{r}
gsea_result <- read_csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/finaldata/GSEA_GO_result_microfuncBP_FC_p05.csv")

gsea_result_number <- gsea_result %>% filter(gsea_result$FDR < 0.5)


# Tidy the results:
gsea_result_Tidy <- gsea_result  %>%
  as_tibble() %>%
  arrange(desc(NES))

gsea_result_Tidy$absNES <- abs(gsea_result_Tidy$NES)

gsea_result_Tidy2 <- gsea_result_Tidy %>% top_n(30, wt=absNES)

# Show in a table - adjust for fgsea by adding dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
gsea_result_Tidy2  %>% 
  #top_n(15, wt= abs(NES)) %>%
  arrange(FDR) %>% 
  DT::datatable()
```

#facet
#Dot plot
```{r}
library(circlize)
colors1=c( "#F8FEBE","#F8FEBE","#F6EAA8","#F5D793","#F4C47E","#F3B069","#F19D54","#F08A3F","#EF762A","#EE6315","#ED5000","#ED5000","#ED5000","#C34C17")
color1 <- colorRampPalette(colors = colors1)(10000)
colors2=c("#F8FEBE","#ED5000")
color2 <- colorRampPalette(colors = colors2)(10)

colors3=colorRamp2(c(0.5,0.75, 1.0, 1.45,1.5,1.55,1.6,1.7,1.8,1.85,1.9,2.0), c( "#F8FEBE","#F9F1B0", "#FBF88B" ,"#FBF487" ,"#FAEC93" ,"#F6D026" ,"#F3C35C" ,"#EDA100", "#EE6600","#F29C4D","#EE5508","#ED5000"))

#color3 <- colorRampPalette(colors = colors3)(10)


# filter P-value
gsea_result_Tidy4 <- gsea_result_Tidy %>% 
  dplyr::select(Goterm, ID, Description, NES, absNES, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 0.5)

# Plot GOplot
gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(30, wt = absNES)


#facet 0 or 1
gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 > 0.2)] <- 0.2


fig_goplot2 <-       gsea_result_Tidy5      %>%
                     ggplot(aes(x = FDR2,
                               y=reorder(Goterm,-FDR2),
                               colour = absNES,
                               fill =absNES,
                               size = pergene*2
                               )) +
                    geom_point()  +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                   # scale_x_continuous(limits=c("0","0.5","1","2")) +
                    scale_x_reverse(limits = c(0.2, 0.011), 
                                    breaks = c(0.2,0.1, 0.05, 0.03, 0.01),
                                    labels = c("≥0.2","0.1","0.05", "0.03", "0.01")
                                    ) +
                    #scale_fill_continuous(trans = "reverse") +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                          limits=c(0.5, 2.5), 
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=F)) +
                    labs(x=expression(paste("False Discovery Rate")), y="GO term", colour="Normalized Enrichment Scores", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), #change the size scale 
                               breaks = c (0.3, 0.4, 0.5, 0.6, 0.7),
                               labels = c( "0.3", "0.4", "0.5", "0.6", "0.7")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=13
                                                 #face="bold",
                                                ) 
                          )
fig_goplot2


#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                      legend.position="bottom"
                      #legend.box="vertical"
                      ) +
                      guides(colour = guide_colourbar(order = 1),
                             fill= "none")


fig_goplot3

ppi=300
par(mar = c(1, 1, 1, 1)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_New_microfuncBP_tstat_top30_",ppi ,"dpi.tiff"), height = 12*ppi, width = 18*ppi, res=ppi)
fig_goplot3
dev.off()
```

#Plot GO pathway
#これつかう　figurefigure
# p-value sort
```{r}
colors1=c( "#F8FEBE","#F8FEBE","#F6EAA8","#F5D793","#F4C47E","#F3B069","#F19D54","#F08A3F","#EF762A","#EE6315","#ED5000","#ED5000","#ED5000","#C34C17")
colors11=c( "#C34C17","#ED5000","#ED5000","#ED5000","#EE6315","#EF762A","#F08A3F","#F19D54","#F3B069","#F4C47E","#F5D793")

color1 <- colorRampPalette(colors = colors11)(10000)
colors2=c("#F8FEBE","#ED5000")
color2 <- colorRampPalette(colors = colors2)(10)

library(circlize)
#colors3=colorRamp2(c(0.5,0.75, 1.0,1.2, 1.4, 1.45,1.5,1.55,1.6,1.65,1.7,1.75,1.8,1.85,1.9,2.0,2.05,2.1,2.15,2.2, 2.25,2.3,2.35,2.4,2.5),color1 )


# filter P-value
gsea_result_Tidy4 <- gsea_result_Tidy %>% 
  dplyr::select(Goterm, ID, Description, NES, absNES, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 0.1)

# Plot GOplot
#gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(30, wt = absNES)
gsea_result_Tidy5 <- gsea_result_Tidy %>% top_n(30, wt=-pvalue)

#facet 0 or 1
gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001
```

```{r}
#Upper  lower alphabet
#gsea_result_Tidy5$Goterm <- str_to_sentence(gsea_result_Tidy5$Goterm, locale = "en")
#個別に修正
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Rna processing (GO:0006396)")] <- "RNA processing (GO:0006396)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Ncrna processing (GO:0034470)")] <- "ncRNA processing (GO:0034470)"

gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="")] <- ""
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="")] <- ""
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="")] <- ""
```


```{r}
fig_goplot2 <-       gsea_result_Tidy5      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05", "≥0.10"),
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", colour="FDR", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), #change the size scale 
                               breaks = c (0.2, 0.3, 0.4, 0.5, 0.6),
                               labels = c( "0.2", "0.3", "0.4", "0.5", "0.6")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=13
                                                 #face="bold",
                                                )
                          )
fig_goplot2

###
#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                      legend.position="right",
                      legend.text = element_text(size=10)#,
                      #legend.box="vertical"
                      ) 


fig_goplot3

ppi=300
par(mar = c(0.1, 0.1, 0.1, 0.1)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_New_microfuncBP_FC_top30_r1.tiff"), height = 11*ppi, width = 11*ppi, res=ppi)
fig_goplot3
dev.off()
```


##########################################
#Cellular Component
###########################################

#GO-GSEA by clusterProfiler gseGO using ttest data (t-statistics)
```{r}
library(clusterProfiler)
library(DOSE)
library(org.Hs.eg.db)
library(ggplot2)
library(dplyr)
library("AnnotationDbi")
library("org.Hs.eg.db")
library("org.EcK12.eg.db")
library("org.EcSakai.eg.db")


data(geneList)
#Prepare Input
# we want the log2 fold change 
#original_gene_list <- res$log2FoldChange
# we want the p-value

ttest_func$ENTREZID = mapIds(org.EcK12.eg.db,
                    keys=  ttest_func$name ,
                    column="ENTREZID",
                    keytype ="SYMBOL",
                    multiVals="first")

#fullgene_list <- res_ttest  %>% filter(!(is.na(res_ttest$p_adjust)))
fullgene_list <- ttest_func %>% filter(!(is.na(ttest_func$stat)))


#original_gene_list <- fullgene_list$p_adjust
original_gene_list <- fullgene_list$stat
# we want the p-value
#original_gene_list <- res$stat
# name the vector
#names(original_gene_list) <- fullgene_list$ENTREZID
names(original_gene_list) <- fullgene_list$name

original_gene_list_df <- as.data.frame(original_gene_list)
original_gene_list_df <- data.frame(fullgene_list$ENTREZID, original_gene_list_df )

#gene_list <- left_join(original_gene_list_df, all.genes.entrez, by = c("rownames.original_gene_list_df." = "ENSEMBL"))
original_gene_list_df <- original_gene_list_df %>% filter(!is.na(original_gene_list_df))

#original_gene_list_df <- original_gene_list_df %>% distinct(fullgene_list.ENTREZID)

## step 0
gene_list_ori <- original_gene_list_df
## Step 1: numeric vector
gene_list <- gene_list_ori[,2]
## Step 2: named vector
names(gene_list) = as.character(gene_list_ori[,1])
## Step 3: decreasing order
gene_list = sort(gene_list, decreasing = TRUE)


#開始時と終了時の差を取る 
start.time<-proc.time() #開始
#GSEA
gse_result <- clusterProfiler::gseGO(geneList     = gene_list ,  # 全遺伝子,
                    OrgDb        = org.EcK12.eg.db,
                    ont          = "CC", # "BP","CC","MF","ALL"から選択
                    #nPerm        = 1000,
                    #nPermSimple = 1000,
                    pvalueCutoff = 1,
                    #qvalueCutoff = 0.15,
                    minGSSize = 10,
                    maxGSSize = 500, 
                    eps = 0, 
                    pAdjustMethod = "fdr", 
                    verbose = TRUE, 
                    seed = 622,
                    by = "fgsea"
                    #pool = T
                    )

gsea_result_df <- as.data.frame(gse_result)
#View(gsea_result_df) 
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

```{r}
gsea_result_df <- as.data.frame(gse_result)
#Upper  lower alphabet
gsea_result_df$Description <- str_to_sentence(gsea_result_df$Description, locale = "en")

#個別に修正
#gsea_result_df$Description[which(gsea_result_df$Description =="Positive t cell selection")] <- "Positive T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Cellular response to type i interferon")] <- "Cellular response to type I interferon"
#gsea_result_df$Description[which(gsea_result_df$Description =="Type i interferon signaling pathway")] <- "Type I interferon signaling pathway"
#gsea_result_df$Description[which(gsea_result_df$Description =="Thymic t cell selection")] <- "Thymic T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Thymic t cell selection")] <- "Thymic T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Positive t cell selection")] <- "Positive T cell selection"
#gsea_result_df$Description[which(gsea_result_df$Description =="Response to type i interferon")] <- "Response to type I interferon"
#gsea_result_df$Description[which(gsea_result_df$Description =="")] <- ""

#GSEAの可視化
library(ggplot2)
library(ggridges)
#ridgeplot(gse_result,showCategory = 8)

write.csv(gsea_result_df, file = "GSEA_GO_result_microfunc_CC.csv", row.names = FALSE)

```

# Gene ratio にてselection
```{r}
N <- dim(gsea_result_df)[1]
#Core enrichmentの数を数える
for (i in 1:N) {
 gsea_result_df$Count[i] <- length(str_extract_all(gsea_result_df$core_enrichment[i], pattern="/", simplify=TRUE)) +1 
}
gsea_result_df$pergene <-  (gsea_result_df$Count/gsea_result_df$setSize)


#Add ID
gsea_result_df$ID <- rownames(gsea_result_df)
rownames(gsea_result_df) <- paste0(gsea_result_df$Description, " (",gsea_result_df$ID, ")" )
gsea_result_df$Goterm <-    rownames(gsea_result_df)  

  
# filter gene percent
gsea_result_df2 <- gsea_result_df %>% 
  dplyr::select(Goterm, ID, Description, NES, pvalue, p.adjust, pergene)  %>% 
  filter(pergene >= 0.05)

#calculate FDR
gsea_result_df2$FDR <- p.adjust(gsea_result_df2$pvalue, method = "fdr")
gsea_result_df3 <- gsea_result_df2 %>% filter(gsea_result_df2$FDR < 0.5)


write.csv(gsea_result_df, file = "GSEA_GO_result_microfuncCC_tstat_ALL.csv", row.names = FALSE)
write.csv(gsea_result_df3, file = "GSEA_GO_result_microfuncCC_tstat_p50.csv", row.names = FALSE)
```

# Show in a tabl
```{r}
gsea_result <- read_csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/finaldata/GSEA_GO_result_microfuncCC_tstat_p50.csv")


# Tidy the results:
gsea_result_Tidy <- gsea_result  %>%
  as_tibble() %>%
  arrange(desc(NES))

gsea_result_Tidy$absNES <- abs(gsea_result_Tidy$NES)

gsea_result_Tidy2 <- gsea_result_Tidy %>% top_n(30, wt=absNES)

# Show in a table - adjust for fgsea by adding dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
gsea_result_Tidy2  %>% 
  #top_n(15, wt= abs(NES)) %>%
  arrange(FDR) %>% 
  DT::datatable()
```

# p-value sort
```{r}
colors1=c( "#F8FEBE","#F8FEBE","#F6EAA8","#F5D793","#F4C47E","#F3B069","#F19D54","#F08A3F","#EF762A","#EE6315","#ED5000","#ED5000","#ED5000","#C34C17")
colors11=c( "#C34C17","#ED5000","#ED5000","#ED5000","#EE6315","#EF762A","#F08A3F","#F19D54","#F3B069","#F4C47E","#F5D793")

color1 <- colorRampPalette(colors = colors11)(10000)
colors2=c("#F8FEBE","#ED5000")
color2 <- colorRampPalette(colors = colors2)(10)

library(circlize)
#colors3=colorRamp2(c(0.5,0.75, 1.0,1.2, 1.4, 1.45,1.5,1.55,1.6,1.65,1.7,1.75,1.8,1.85,1.9,2.0,2.05,2.1,2.15,2.2, 2.25,2.3,2.35,2.4,2.5),color1 )


# filter P-value
gsea_result_Tidy4 <- gsea_result_Tidy %>% 
  dplyr::select(Goterm, ID, Description, NES, absNES, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 0.1)

# Plot GOplot
#gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(30, wt = absNES)
gsea_result_Tidy5 <- gsea_result_Tidy %>% top_n(30, wt=-pvalue)

#facet 0 or 1
gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001
```

```{r}
#Upper  lower alphabet
#gsea_result_Tidy5$Goterm <- str_to_sentence(gsea_result_Tidy5$Goterm, locale = "en")
#個別に修正
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Nadh dehydrogenase complex (GO:0030964)")] <- "NADH dehydrogenase complex (GO:0030964)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Respiratory chain complex i (GO:0045271)")] <- "Respiratory chain complex I (GO:0045271)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Plasma membrane respiratory chain complex i (GO:0045272)")] <- "Plasma membrane respiratory chain complex I (GO:0045272"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="")] <- ""
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="")] <- ""
```


```{r}
fig_goplot2 <-       gsea_result_Tidy5      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05", "≥0.10"),
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", colour="FDR", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), #change the size scale 
                               breaks = c (0.2, 0.3, 0.4, 0.5, 0.6),
                               labels = c( "0.2", "0.3", "0.4", "0.5", "0.6")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=13
                                                 #face="bold",
                                                )
                          )
fig_goplot2

###
#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                      legend.position="right",
                      legend.text = element_text(size=10)#,
                      #legend.box="vertical"
                      ) 


fig_goplot3

ppi=300
par(mar = c(0.1, 0.1, 0.1, 0.1)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_New_microfuncCC_tstat_top30_",ppi ,"dpi.tiff"), height = 10*ppi, width = 10.5*ppi, res=ppi)
fig_goplot3
dev.off()
```


##########################################
#Molecular function_micorbiome
###########################################

#GO-GSEA by clusterProfiler gseGO using ttest data (t-statistics)
```{r}
library(clusterProfiler)
library(DOSE)
library(org.Hs.eg.db)
library(ggplot2)
library(dplyr)
library("AnnotationDbi")
library("org.Hs.eg.db")
library("org.EcK12.eg.db")
library("org.EcSakai.eg.db")

#開始時と終了時の差を取る 
start.time<-proc.time() #開始
#GSEA
gse_result <- clusterProfiler::gseGO(geneList     = gene_list ,  # 全遺伝子,
                    OrgDb        = org.EcK12.eg.db,
                    ont          = "MF", # "BP","CC","MF","ALL"から選択
                    #nPerm        = 1000,
                    #nPermSimple = 1000,
                    pvalueCutoff = 1,
                    #qvalueCutoff = 0.15,
                    minGSSize = 10,
                    maxGSSize = 500, 
                    eps = 0, 
                    pAdjustMethod = "fdr", 
                    verbose = TRUE, 
                    seed = 622,
                    by = "fgsea"
                    #pool = T
                    )

gsea_result_df <- as.data.frame(gse_result)
#View(gsea_result_df) 
end.time<-proc.time() #終了
(end.time-start.time) #経過部分が開始から終了までの時間に該当
```

```{r}
gsea_result_df <- as.data.frame(gse_result)
#Upper  lower alphabet
gsea_result_df$Description <- str_to_sentence(gsea_result_df$Description, locale = "en")
library(ggplot2)
library(ggridges)
#ridgeplot(gse_result,showCategory = 8)

write.csv(gsea_result_df, file = "GSEA_GO_result_microfunc_MF.csv", row.names = FALSE)

```

# Gene ratio にてselection
```{r}
N <- dim(gsea_result_df)[1]
#Core enrichmentの数を数える
for (i in 1:N) {
 gsea_result_df$Count[i] <- length(str_extract_all(gsea_result_df$core_enrichment[i], pattern="/", simplify=TRUE)) +1 
}
gsea_result_df$pergene <-  (gsea_result_df$Count/gsea_result_df$setSize)


#Add ID
gsea_result_df$ID <- rownames(gsea_result_df)
rownames(gsea_result_df) <- paste0(gsea_result_df$Description, " (",gsea_result_df$ID, ")" )
gsea_result_df$Goterm <-    rownames(gsea_result_df)  

  
# filter gene percent
gsea_result_df2 <- gsea_result_df %>% 
  dplyr::select(Goterm, ID, Description, NES, pvalue, p.adjust, pergene)  %>% 
  filter(pergene >= 0.05)

#calculate FDR
gsea_result_df2$FDR <- p.adjust(gsea_result_df2$pvalue, method = "fdr")
gsea_result_df3 <- gsea_result_df2 %>% filter(gsea_result_df2$FDR < 0.5)


write.csv(gsea_result_df, file = "GSEA_GO_result_microfuncMF_tstat_ALL.csv", row.names = FALSE)
write.csv(gsea_result_df3, file = "GSEA_GO_result_microfuncMF_tstat_p50.csv", row.names = FALSE)
```

# Show in a tabl
```{r}
gsea_result <- read_csv("/Users/mf65/Dropbox/MGH/MARC/MARC10_dual/Analysis_R_m10/finaldata/GSEA_GO_result_microfuncMF_tstat_p50.csv")


# Tidy the results:
gsea_result_Tidy <- gsea_result  %>%
  as_tibble() %>%
  arrange(desc(NES))

gsea_result_Tidy$absNES <- abs(gsea_result_Tidy$NES)

gsea_result_Tidy2 <- gsea_result_Tidy %>% top_n(30, wt=absNES)

# Show in a table - adjust for fgsea by adding dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
gsea_result_Tidy2  %>% 
  #top_n(15, wt= abs(NES)) %>%
  arrange(FDR) %>% 
  DT::datatable()
```


# p-value sort
```{r}
colors1=c( "#F8FEBE","#F8FEBE","#F6EAA8","#F5D793","#F4C47E","#F3B069","#F19D54","#F08A3F","#EF762A","#EE6315","#ED5000","#ED5000","#ED5000","#C34C17")
colors11=c( "#C34C17","#ED5000","#ED5000","#ED5000","#EE6315","#EF762A","#F08A3F","#F19D54","#F3B069","#F4C47E","#F5D793")

color1 <- colorRampPalette(colors = colors11)(10000)
colors2=c("#F8FEBE","#ED5000")
color2 <- colorRampPalette(colors = colors2)(10)

library(circlize)
#colors3=colorRamp2(c(0.5,0.75, 1.0,1.2, 1.4, 1.45,1.5,1.55,1.6,1.65,1.7,1.75,1.8,1.85,1.9,2.0,2.05,2.1,2.15,2.2, 2.25,2.3,2.35,2.4,2.5),color1 )


# filter P-value
gsea_result_Tidy4 <- gsea_result_Tidy %>% 
  dplyr::select(Goterm, ID, Description, NES, absNES, pvalue, p.adjust, pergene, FDR)  %>% 
  filter(FDR < 1)

# Plot GOplot
#gsea_result_Tidy5 <- gsea_result_Tidy4 %>% top_n(30, wt = absNES)
gsea_result_Tidy5 <- gsea_result_Tidy %>% top_n(30, wt=-pvalue)

#facet 0 or 1
gsea_result_Tidy5$status <- ifelse(gsea_result_Tidy5$NES > 0, "Up-regulated", "Down-regulated")

#FDR round
gsea_result_Tidy5$FDR2 <- round(gsea_result_Tidy5$FDR, 3)
gsea_result_Tidy5$FDR2[which(gsea_result_Tidy5$FDR2 ==0.000)] <- 0.001
```

```{r}
#Upper  lower alphabet
#gsea_result_Tidy5$Goterm <- str_to_sentence(gsea_result_Tidy5$Goterm, locale = "en")
#個別に修正

gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Catalytic activity, acting on a trna (GO:0140101)")] <- "Catalytic activity, acting on a tRNA (GO:0140101)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Catalytic activity, acting on rna (GO:0140098)")] <- "Catalytic activity, acting on RNA (GO:0140098)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Trna binding (GO:0000049)")] <- "tRNA binding (GO:0000049)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Rrna binding (GO:0019843)")] <- "rRNA binding (GO:0019843)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Oxidoreductase activity, acting on nad(p)h (GO:0016651)")] <- "Oxidoreductase activity, acting on NAD(P)H (GO:0016651)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Oxidoreductase activity, acting on the ch-ch group of donors, nad or nadp as acceptor (GO:0016628)")] <- "Oxidoreductase activity, acting on the CH-CH group of donors, NAD or NADP as acceptor (GO:0016628)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Oxidoreductase activity, acting on nad(p)h, quinone or similar compound as acceptor (GO:0016655)")] <- "Oxidoreductase activity, acting on NAD(P)H, quinone or similar compound as acceptor (GO:0016655)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Nadh dehydrogenase (ubiquinone) activity (GO:0008137)")] <- "NADH dehydrogenase (ubiquinone) activity (GO:0008137)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Nadh dehydrogenase (quinone) activity (GO:0050136)")] <- "NADH dehydrogenase (quinone) activity (GO:0050136)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Trna methyltransferase activity (GO:0008175)")] <- "tRNA methyltransferase activity (GO:0008175)"
gsea_result_Tidy5$Goterm[which(gsea_result_Tidy5$Goterm =="Nadh dehydrogenase activity (GO:0003954)")] <- "NADH dehydrogenase activity (GO:0003954)"
```


```{r}
fig_goplot2 <-       gsea_result_Tidy5      %>%
                    ggplot(aes(x = absNES,
                               y=reorder(Goterm,absNES),
                               colour = FDR2,
                               size = pergene)) +
                    geom_point() +
                    facet_grid( ~ status) +
                    expand_limits(  size= c (0.3, 0.8)) + 
                    #scale_x_discrete(limits=c("2","1","0.5")) +
                    scale_x_continuous(limits = c(1, 2.5), 
                                    breaks = c(1,2, 2.5),
                                    labels = c("≤1", "2",  "2.5")
                                    ) +
                    scale_colour_gradientn(colours = color1 , 
                                           space = "Lab",
                                           limits=c(0.001,0.1), 
                                           breaks = c (0.001, 0.05, 0.1),
                                           labels = c( "≤0.001", "0.05", "≥0.10"),
                                           oob = scales::squish,
                                           na.value = "#C6DBEF" , 
                                           guide=guide_colorbar(reverse=T)) +
                    labs(x=expression(paste("Absolute normalized enrichment scores")), y="GO term", colour="FDR", size="Gene ratio") + 
                    scale_size_continuous(range = c(3, 10), #change the size scale 
                               breaks = c (0.2, 0.3, 0.4, 0.5, 0.6),
                               labels = c( "0.2", "0.3", "0.4", "0.5", "0.6")) +
                    #theme_minimal() + 
                    theme(strip.background = element_rect(colour="black", 
                                                          fill="white", 
                                                          size=1.5, 
                                                          linetype="solid"),
                          strip.text.x = element_text(size = 15, color = "black", face = "bold"),
                          axis.text.x = element_text(size=13
                                                     #face="bold", 
                                                     ),
                          axis.text.y = element_text(size=13 
                                               ), 
                          axis.title.x = element_text(size=13
                                                #face="bold", 
                                                ), 
                          axis.title.y = element_text( size=13
                                                 #face="bold",
                                                )
                          )
fig_goplot2

###
#spacing between each facet
fig_goplot3  <- fig_goplot2 + 
                theme(panel.spacing.x=unit(2, "lines"),
                      legend.position="right",
                      legend.text = element_text(size=10)#,
                      #legend.box="vertical"
                      ) 


fig_goplot3

ppi=300
par(mar = c(0.1, 0.1, 0.1, 0.1)); # c(lower ,left, upper, right)
tiff(paste0 ("Goplot_New_microfuncMF_tstat_top30_",ppi ,"dpi.tiff"), height = 12*ppi, width = 13*ppi, res=ppi)
fig_goplot3
dev.off()
```
